{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: reshape\n",
      "\n",
      "Attaching package: ‘reshape’\n",
      "\n",
      "The following object is masked from ‘package:Matrix’:\n",
      "\n",
      "    expand\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    melt\n",
      "\n",
      "Loading required package: scales\n",
      "Loading required package: ggpubr\n",
      "Loading required package: magrittr\n",
      "\n",
      "Attaching package: ‘ggpubr’\n",
      "\n",
      "The following object is masked from ‘package:cowplot’:\n",
      "\n",
      "    get_legend\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'base_Rdata_name'</li>\n",
       "\t<li>'cell_assignment_gps'</li>\n",
       "\t<li>'HVG'</li>\n",
       "\t<li>'patient_type'</li>\n",
       "\t<li>'sample_list'</li>\n",
       "\t<li>'t2g'</li>\n",
       "\t<li>'taggene_hk'</li>\n",
       "\t<li>'taggene_mt'</li>\n",
       "\t<li>'tissue_type'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'base\\_Rdata\\_name'\n",
       "\\item 'cell\\_assignment\\_gps'\n",
       "\\item 'HVG'\n",
       "\\item 'patient\\_type'\n",
       "\\item 'sample\\_list'\n",
       "\\item 't2g'\n",
       "\\item 'taggene\\_hk'\n",
       "\\item 'taggene\\_mt'\n",
       "\\item 'tissue\\_type'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'base_Rdata_name'\n",
       "2. 'cell_assignment_gps'\n",
       "3. 'HVG'\n",
       "4. 'patient_type'\n",
       "5. 'sample_list'\n",
       "6. 't2g'\n",
       "7. 'taggene_hk'\n",
       "8. 'taggene_mt'\n",
       "9. 'tissue_type'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"base_Rdata_name\"     \"cell_assignment_gps\" \"HVG\"                \n",
       "[4] \"patient_type\"        \"sample_list\"         \"t2g\"                \n",
       "[7] \"taggene_hk\"          \"taggene_mt\"          \"tissue_type\"        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "suppressMessages(require(ggplot2))\n",
    "suppressMessages(require(data.table))\n",
    "suppressMessages(require(gplots))\n",
    "suppressMessages(require(Seurat))\n",
    "suppressMessages(require(Matrix))\n",
    "require(reshape)\n",
    "require(scales)\n",
    "require(\"ggpubr\")\n",
    "source(\"VDJ_ploting_suite.R\")\n",
    "source(\"VDJ_function_pack.R\")\n",
    "\n",
    "load(\"180607_misc.Rdata\")\n",
    "base_Rdata_name=\"180607_RUVscale\"\n",
    "lsnofun()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "GWAS_gene=c(\"APOBEC3G\",\"ARHGAP30\",\"ARPC2\",\"ATF4\",\"ATG16L1\",\"BRE\",\"BSN\",\"C10orf58\",\"C11orf9\",\"C1orf53\",\"C2orf74\",\"CALM3\",\"CARD11\",\"CARD9\",\"CCDC116\",\"CCDC88B\",\"CCL11\",\"CCL13\",\"CCL2\",\"CCR6\",\"CD226\",\"CD244\",\"CD40\",\"MMP9\",\"PLTP\",\"CD48\",\"CD5\",\"CD6\",\"CEBPB\",\"CEBPG\",\"CEP250\",\"CISD1\",\"CNTF\",\"CPEB4\",\"CREB5\",\"CREM\",\"CRTC3\",\"CSF2\",\"CTDSP1\",\"CTSW\",\"CTSZ\",\"CXCL1\",\"CXCL2\",\"CXCL3\",\"CXCL5\",\"CXCL6\",\"CXCR1\",\"CXCR2\",\"CXCR5\",\"DAP\",\"DBP\",\"DCLRE1B\",\"DLD\",\"DNMT3B\",\"DOK3\",\"EIF3C\",\"EPO\",\"ERAP1\",\"ERAP2\",\"F11R\",\"FADS1\",\"FADS2\",\"FAIM3\",\"FAM55A\",\"FAM55D\",\"FASLG\",\"FCGR2A\",\"FCGR2B\",\"FCGR3A\",\"FCGR3B\",\"FCRLA\",\"FLRT1\",\"FOS\",\"FOSL1\",\"FOSL2\",\"FUT2\",\"FYN\",\"GALC\",\"GART\",\"GNA12\",\"GPR18\",\"GPR183\",\"GPR35\",\"GPR65\",\"GPX1\",\"GPX4\",\"GSDMA\",\"GSDMB\",\"HCK\",\"HLA-C\",\"HLA-DQA1\",\"HLA-DQB1\",\"HLA-DRA\",\"HLA-DRB1\",\"HMHA1\",\"HNF4A\",\"HSPA6\",\"ICAM1\",\"ICOSLG\",\"IFIH1\",\"IFNAR1\",\"IFNAR2\",\"IFNG\",\"IFNGR2\",\"IKZF1\",\"IKZF3\",\"IL10\",\"IL10RB\",\"IL12B\",\"IL12RB2\",\"IL13\",\"IL15RA\",\"IL18R1\",\"IL18RAP\",\"IL19\",\"IL1R1\",\"IL1R2\",\"IL1RL1\",\"IL1RL2\",\"IL2\",\"IL20\",\"IL21\",\"IL22\",\"IL23R\",\"IL24\",\"IL26\",\"IL27\",\"IL2RA\",\"IL3\",\"IL31RA\",\"IL4\",\"IL5\",\"IL6ST\",\"IL8\",\"INPP5D\",\"\",\"INPP5E\",\"IP6K1\",\"IP6K2\",\"IPMK\",\"IRF1\",\"IRF5\",\"IRF8\",\"IRGM\",\"ITGAL\",\"ITIH4\",\"ITLN1\",\"ITPKA\",\"IZUMO1\",\"JAK2\",\"JAZF1\",\"JRKL\",\"KIAA1841\",\"KIF21B\",\"KIR2DL1\",\"LACC1\",\"LGALS9\",\"LIF\",\"LILRB4\",\"LIME1\",\"LITAF\",\"LNPEP\",\"LOH12CR1\",\"LPXN\",\"LRRK2\",\"LSP1\",\"MAML2\",\"MANBA\",\"MAP3K8\",\"MAPK1\",\"MAPKAPK2\",\"MICB\",\"MLH3\",\"MMEL1\",\"MST1\",\"MST1R\",\"MSTO1\",\"MTMR3\",\"MUC19\",\"NDFIP1\",\"NDUFAF1\",\"NFIL3\",\"NFKB1\",\"NFKBIL1\",\"NKX2-3\",\"NLRP2\",\"NLRP7\",\"NOD2\",\"NOS2\",\"NUPR1\",\"NUSAP1\",\"ORMDL3\",\"OSM\",\"PDLIM4\",\"PF4\",\"PF4V1\",\"PFKFB4\",\"PHACTR2\",\"PIGR\",\"PLCH2\",\"PLCL1\",\"PMPCA\",\"PNKD\",\"PPAN-P2RY11\",\"PRKCB\",\"PRKCD\",\"PROCR\",\"PSORS1C1\",\"PTGDR2\",\"PTGER4\",\"PTPN22\",\"RABEP2\",\"RASGRP1\",\"RASSF5\",\"REL\",\"RELA\",\"REV3L\",\"RFTN2\",\"RIMBP3\",\"RIPK2\",\"RIT1\",\"RMI2\",\"RNASET2\",\"RORC\",\"RPS6KA2\",\"RPS6KA4\",\"RPS6KB1\",\"SDCCAG3\",\"SLAMF1\",\"SLAMF7\",\"SLC10A4\",\"SLC11A1\",\"SLC22A4\",\"SLC22A5\",\"SLC2A4RG\",\"SLC9A3\",\"SMAD3\",\"SMAD7\",\"SMURF1\",\"SNX32\",\"SOCS1\",\"SP140\",\"SPHK2\",\"SPRED1\",\"SPRED2\",\"SPRY4\",\"STAT1\",\"STAT3\",\"STAT4\",\"STAT5A\",\"STAT5B\",\"SULT1A1\",\"SULT1A2\",\"TAB1\",\"TAGAP\",\"TEC\",\"TMBIM1\",\"TMEM50B\",\"TNC\",\"TNFAIP3\",\"TNFRSF14\",\"TNFRSF18\",\"TNFRSF4\",\"TNFRSF6B\",\"TNFRSF9\",\"TNFSF15\",\"TNFSF18\",\"TNFSF8\",\"TNIP1\",\"TNNI2\",\"TNPO3\",\"TRAF3IP2\",\"TRIB1\",\"TRPT1\",\"TSPAN14\",\"TSPAN33\",\"TTYH3\",\"TUBD1\",\"TXK\",\"TYK2\",\"UBE2L3\",\"UBQLN4\",\"UCN\",\"UCN2\",\"UQCC\",\"USF1\",\"USP4\",\"VDR\",\"YDJC\",\"ZFP36L1\",\"ZFP90\",\"ZGPAT\",\"ZNF300P1\",\"ZNF831\",\"ZPBP\",\"ZPBP2\")\n",
    "GWAS_gene=toupper(GWAS_gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## T Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_Tcell_R2.Rds\")\n",
    "#load(\"abgdTcell_labeling.Rdata\")\n",
    "tag_celltype=\"Tcell\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#cell type assignment\n",
    "celltype_assignment_df=data.frame(cluster=c(\"17,16\",\"8,2\",\"13,5\",\"6,0\",\"7,10\",\"11,1\",\"3\",\"18,12,9,19\",\"21,20\",\"4,14,15\"),\n",
    "                                  celltype=c(\"Naive_B\",\"Memory_B\",\"Plasma\",\"NKT\",\"CD8\",\"CD4\",\"gdT\",\"NK\",\"unknown\",\"M_phi\"),\n",
    "                                  stringsAsFactors=F)\n",
    "\n",
    "celltype_assignment_expanded=cellTypeDF_processing(celltype_assignment_df)\n",
    "\n",
    "celltype_assignment=gp_name_replacing(as.character(Exp_Seurat@meta.data$global_gps),celltype_assignment_expanded$cluster,celltype_assignment_expanded$celltype)\n",
    "subcelltype_assignment=gp_name_replacing(as.character(Exp_Seurat@meta.data$global_gps),celltype_assignment_expanded$cluster,celltype_assignment_expanded$subcelltype)\n",
    "\n",
    "#cell cyle score\n",
    "#Exp_Seurat=cellcycle_assigning(Exp_Seurat,cycle_gene_list=\"~/Analysis/genomeFiles/regev_lab_cell_cycle_genes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#cell type assignment\n",
    "Tcell_assignment_df=data.frame(cluster=c(\"12,9,5,15\",\"6,16\",\"11\",\"1,0,10,7\",\"14\",\"4,17,3,8,2,13\"),\n",
    "                                  celltype=c(\"CD8_TEM_effector\",\"CD8\",\"CD8_TCM\",\"NKT\",\"gdT\",\"CD4\"),\n",
    "                                  stringsAsFactors=F)\n",
    "Tcell_assignment_df=data.frame(cluster=c(\"12,9,5,15,6,16,11\",\"1,0,10,7\",\"14\",\"4,17,3,8,2,13\"),\n",
    "                                  celltype=c(\"CD8\",\"NKT\",\"gdT\",\"CD4\"),\n",
    "                                  stringsAsFactors=F)\n",
    "Tcell_assignment_expanded=cellTypeDF_processing(Tcell_assignment_df)\n",
    "Tcell_assignment=gp_name_replacing(as.character(Exp_Seurat@ident),Tcell_assignment_expanded$cluster,Tcell_assignment_expanded$celltype)\n",
    "Tcellsubcelltype_assignment=gp_name_replacing(as.character(Exp_Seurat@ident),Tcell_assignment_expanded$cluster,Tcell_assignment_expanded$subcelltype)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Tcellsubcelltype_assignment[Exp_Seurat@meta.data$disease_assignment==1]=paste0(Tcellsubcelltype_assignment[Exp_Seurat@meta.data$disease_assignment==1],\"_healthy\")\n",
    "Tcellsubcelltype_assignment[Exp_Seurat@meta.data$disease_assignment==2]=paste0(Tcellsubcelltype_assignment[Exp_Seurat@meta.data$disease_assignment==2],\"_Disease\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### cell percentage of each cell type in tissue/patient/disease_status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample_cell_num=readRDS(file=\"sample_cell_num.rds\")\n",
    "\n",
    "tmp=table(Tcellsubcelltype_assignment,Exp_Seurat@meta.data$sample_gp)\n",
    "tmp_rownames=rownames(tmp)\n",
    "colnames(tmp)=sample_list[as.numeric(colnames(tmp))]\n",
    "tmp_percent_by_celltype=tmp\n",
    "tmp_percent_by_all=tmp\n",
    "csum=colSums(tmp)\n",
    "for(i in 1:nrow(tmp)){\n",
    "    tmp_percent_by_celltype[i,]=tmp[i,]/csum\n",
    "    tmp_percent_by_all[i,]=tmp[i,]/sample_cell_num[match(colnames(tmp),names(sample_cell_num))]\n",
    "}\n",
    "tmp=data.table(as.data.frame.matrix(tmp))\n",
    "tmp_percent_by_celltype=data.table(as.data.frame.matrix(tmp_percent_by_celltype))\n",
    "tmp_percent_by_all=data.table(as.data.frame.matrix(tmp_percent_by_all))\n",
    "tmp$majorcelltype=tmp_rownames\n",
    "tmp_percent_by_celltype$majorcelltype=tmp_rownames\n",
    "tmp_percent_by_all$majorcelltype=tmp_rownames\n",
    "\n",
    "tmp=melt(tmp, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"count\")\n",
    "tmp_percent_by_celltype=melt(tmp_percent_by_celltype, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp_percent_by_celltype),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"percent_by_celltype\")\n",
    "tmp_percent_by_all=melt(tmp_percent_by_all, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp_percent_by_all),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"percent_by_all\")\n",
    "tmp$percent_by_celltype=tmp_percent_by_celltype$percent_by_celltype\n",
    "tmp$percent_by_all=tmp_percent_by_all$percent_by_all\n",
    "tmp$patient=gsub(\"_.*\",\"\",tmp$batch)\n",
    "tmp$tissue=gsub(\"^.*_\",\"\",tmp$batch)\n",
    "tmp$disease_status=gp_name_replacing(substr(tmp$batch, 1, 1),c(\"C\",\"U\"),c(\"Healthy\",\"IBD\"))\n",
    "tmp=data.frame(tmp)\n",
    "\n",
    "Group_type=unique(Tcellsubcelltype_assignment)\n",
    "Group_type=Group_type[gtools::mixedorder(Group_type)]\n",
    "\n",
    "pdf(paste0(tag_celltype,\" subtype distribution.pdf\"),width=12,height=5)\n",
    "\n",
    "coltouse=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(Group_type))\n",
    "#options(repr.plot.width=12, repr.plot.height=5)\n",
    "ggplot() +\n",
    "geom_bar(aes(y = count, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"count\") + labs(y = \"count of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = count, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"stack\") +\n",
    "scale_y_continuous() +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"count\") + labs(y = \"count of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_celltype, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by cell type\") + labs(y = paste0(\"pecentage of cells in given batch's \",tag_celltype),fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_all, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by all cells\") + labs(y = \"pecentage of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_all, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"stack\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by all cells\") + labs(y = \"pecentage of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "dev.off()\n",
    "\n",
    "# define the summary function\n",
    "f <- function(x) {\n",
    "  r <- quantile(x, probs = c(0.1, 0.25, 0.5, 0.75, 0.9))\n",
    "  names(r) <- c(\"ymin\", \"lower\", \"middle\", \"upper\", \"ymax\")\n",
    "  r\n",
    "}\n",
    "\n",
    "o <- function(x) {\n",
    "  subset(x, x < quantile(x)[2] | quantile(x)[4] < x)\n",
    "}\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(3,\"Set1\")\n",
    "\n",
    "options(warn=-1) #turn off warning\n",
    "pdf(paste0(\"healthy vs IBD in \",tag_celltype,\" subtype.pdf\"),width=25,height=12)\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=count, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"count of cells in given batch\") + ggtitle(\"count\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=percent_by_celltype, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"pecentage of cells in given batch\") + ggtitle(\"percent by cell type\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=percent_by_all, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"pecentage of cells in given batch\") + ggtitle(\"percent by all cells\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "dev.off()\n",
    "options(warn=0) #turn on warning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### gp gene exp heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "#contrust matrix, each row is a state\n",
    "ExpressionMat_by_gp=cbind(Tcellsubcelltype_assignment,data.frame(2^t(Exp_Seurat@data))) #\n",
    "colnames(ExpressionMat_by_gp)[1]=\"states\"\n",
    "tagV=c(\"TRAV\",\"TRAC\",\"TRBV\",\"TRBC\",\"TRGV\",\"TRGC\",\"TRDV\",\"TRDC\",\n",
    "       \"IGHV\",\"IGHG\",\"IGHA\",\"IGHD\",\"IGHM\",\"IGHE\",\"IGLV\",\"IGLL\",\"IGKV\",\"IGLC\",\"IGKC\")\n",
    "togetrid=c(\"IGHMBP2\")\n",
    "for(i in 1:length(tagV)){\n",
    "    tagG=grep(paste0(\"^\",tagV[i]),colnames(ExpressionMat_by_gp))\n",
    "    tagG=setdiff(tagG,togetrid)\n",
    "    if(length(tagG)==0){\n",
    "        next\n",
    "    }\n",
    "    tmp=ExpressionMat_by_gp[,tagG]\n",
    "    if(!is.null(ncol(tmp))){\n",
    "        tmp=rowSums(tmp)\n",
    "    }\n",
    "    ExpressionMat_by_gp=cbind(ExpressionMat_by_gp,tmp)\n",
    "    colnames(ExpressionMat_by_gp)[ncol(ExpressionMat_by_gp)]=tagV[i]\n",
    "}\n",
    "ExpressionMat_by_gp=data.table(ExpressionMat_by_gp)\n",
    "ExpressionMat_by_gp=ExpressionMat_by_gp[,lapply(.SD, mean), by=states]\n",
    "ExpressionMat_by_gp=ExpressionMat_by_gp[order(ExpressionMat_by_gp$states)]\n",
    "rn=ExpressionMat_by_gp$states\n",
    "ExpressionMat_by_gp[,states:=NULL]\n",
    "ExpressionMat_by_gp=data.frame(ExpressionMat_by_gp)\n",
    "colnames(ExpressionMat_by_gp)=gsub(\"\\\\.\",\"-\",colnames(ExpressionMat_by_gp))\n",
    "#ExpressionMat_by_gp=log2(ExpressionMat_by_gp)\n",
    "rownames(ExpressionMat_by_gp)=rn\n",
    "write.table(t(ExpressionMat_by_gp),paste0(\"mediam_gene_expression_\",tag_celltype,\".tsv\"),sep=\"\\t\",col.names=T,row.names=T,quote=F)\n",
    "saveRDS(ExpressionMat_by_gp,file=paste0(\"ExpressionMat_by_gp_\",tag_celltype,\".Rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "top_g=readDICE(DICE_path=\"/home/ahe/Analysis/201801_JohnVDJ/data/DICE_database\")\n",
    "top_g=top_g[,c(\"CD4_NAIVE\",\"CD4_STIM\",\"CD8_NAIVE\",\"CD8_STIM\",\"TFH\",\"TH17\",\"TH1\",\"THSTAR\",\"TREG_MEM\",\"TREG_NAIVE\")]\n",
    "\n",
    "tmp=unique(unlist(top_g))\n",
    "toplot=ExpressionMat_by_gp[,colnames(ExpressionMat_by_gp) %in% tmp]\n",
    "\n",
    "ColforSubCelltype=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(ncol(top_g))\n",
    "\n",
    "ColSideColors=matrix(\"white\",ncol(toplot),ncol(top_g))\n",
    "colnames(ColSideColors)=colnames(top_g)\n",
    "for(i in 1:ncol(top_g)){\n",
    "    tagG=which(colnames(toplot) %in% top_g[,i])\n",
    "    ColSideColors[tagG,i]=ColforSubCelltype[i]\n",
    "}\n",
    "\n",
    "#ColSideColors can be a data.frame with each col indicate one color assignment\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "pdf(file=paste0(tag_celltype,\"_DEGbycluster_heatmap_V2.pdf\"),width=10,height=10)\n",
    "heatmap3::heatmap3(scale(toplot),col=colorRampPalette(col2)(50),useRaster=F,ColSideColors=ColSideColors,\n",
    "                   labCol=F,showColDendro=F,showRowDendro=F,scale='none')\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ExpressionMat_by_gp=readRDS(paste0(\"ExpressionMat_by_gp_\",tag_celltype,\".Rds\"))\n",
    "\n",
    "#match gp with cell type\n",
    "#matching_tbl=table(as.numeric(Exp_Seurat@meta.data$res.1),celltype_assignment)\n",
    "#ct_to_attach=colnames(matching_tbl)[apply(matching_tbl,1,which.max)]\n",
    "#rownames(ExpressionMat_by_gp)=paste0(rownames(ExpressionMat_by_gp),\"_\",ct_to_attach)\n",
    "\n",
    "\n",
    "#T cell\n",
    "taggene=c(\"Ahr\",\"Bach2\",\"Batf\",\"Bcl2\",\"Bcl6\",\"Bhlhe40\",\"Ccr7\",\"Cd101\",\"Cd244\",\"Cd28\",\"Cd3e\",\"Cd3g\",\"CD247\",\"CD4\",\"SPN\",\n",
    "          \"Cd69\",\"Cd8a\",\"Cd8b\",\"Crtam\",\"Ctla4\",\"Cx3cr1\",\"Cxcl13\",\"Cxcr3\",\"Cxcr5\",\"Cxcr6\",\"Dusp2\",\"Dusp5\",\"Dusp6\",\"Eomes\",\n",
    "          \"Ezh2\",\"Faslg\",\"Foxo1\",\"Gzma\",\"Gzmb\",\"Havcr2\",\"Icos\",\"Id2\",\"Id3\",\"Ifng\",\"Ifngr1\",\"Il10\",\"Il2\",\"Il21r\",\"Il2ra\",\n",
    "          \"Il2rb\",\"Il4r\",\"Il7r\",\"Irf4\",\"Itga1\",\"Itga4\",\"Itgae\",\"Itgb7\",\"Itk\",\"Junb\",\"Jund\",\"Klf2\",\"Klf3\",\"Klf6\",\"Klre1\",\n",
    "          \"Klrg1\",\"Lag3\",\"Lat\",\"Lef1\",\"Litaf\",\"Malat1\",\"Myc\",\"Neurl3\",\"Notch1\",\"Nr4a1\",\"Nr4a2\",\"Nr4a3\",\"P2rx7\",\"Pdcd1\",\n",
    "          \"Pdcd4\",\"Pnrc1\",\"Prdm1\",\"Rbm39\",\"Rgs1\",\"Runx3\",\"S1pr1\",\"S1pr4\",\"S1pr5\",\"Sell\",\"Slamf6\",\"Socs3\",\"Stk17b\",\"Tbx21\",\n",
    "          \"Tcf7\",\"TRAV\",\"TRBV\",\"Tgfbr1\",\"Tgfbr2\",\"Tigit\",\"Tnf\",\"Tnfaip3\",\"Tox\",\"Ubc\",\"Vps37b\",\"Xist\",\"Zeb2\",\"Zfp683\",\n",
    "          \"ABCB1\",\"AHR\",\"Areg\",\"Asb2\",\"B3GAT1\",\"BATF\",\"Bcl6\",\"Ccl20\",\"CCL5\",\"Ccr1\",\"CCR10\",\"CCR3\",\"CCR4\",\"CCR5\",\"CCR6\",\n",
    "          \"CCR7\",\"CCR8\",\"CD27\",\"HAVCR2\",\"CD38\",\"ENTPD1\",\"Cd3e\",\"Cd3g\",\"CD247\",\"CD4\",\"CD40LG\",\"Cd69\",\"NT5E\",\"CD84\",\"Cd8a\",\n",
    "          \"Cd8b\",\"PTGDR2\",\"CTLA4\",\"CTSH\",\"Cx3cr1\",\"CXCR3\",\"CXCR5\",\"Cxcr6\",\"EBI3\",\"Eomes\",\"Ezh2\",\"FAS\",\"Foxo1\",\"Foxp3\",\n",
    "          \"Gata3\",\"CSF2\",\"Gzmb\",\"GZMK\",\"HAVCR1\",\"Havcr2\",\"HIF1A\",\"ICOS\",\"Id2\",\"Id3\",\"Ifng\",\"IFNGR1\",\"IKZF1\",\"Ikzf2\",\"IL10\",\n",
    "          \"IL12A\",\"Il12B\",\"Il12rb2\",\"Il13\",\"IL15\",\"Il17A\",\"Il17B\",\"IL17F\",\"IL17RA\",\"IL17re\",\"Il18\",\"IL18R1\",\"Il18RAP\",\n",
    "          \"IL1r1\",\"IL1Rl1\",\"IL2\",\"Il21\",\"Il21r\",\"IL22\",\"IL23A\",\"IL23R\",\"IL25\",\"IL26\",\"IL27\",\"IL27RA\",\"Il2RA\",\"IL3\",\"IL31\",\n",
    "          \"EBI3\",\"IL4\",\"IL4i1\",\"Il5\",\"IL6\",\"IL6R\",\"IL7R\",\"IL9\",\"Il9r\",\"IRF4\",\"IRF8\",\"Itga1\",\"Itga4\",\"Itgae\",\"Itgb7\",\"Klf2\",\n",
    "          \"Klf3\",\"Klf6\",\"KLRB1\",\"Klrg1\",\"Lag3\",\"LAMP1\",\"Lat\",\"Lef1\",\"Litaf\",\"LTA\",\"LTBR\",\"LTK\",\"MAF\",\"Malat1\",\"Myc\",\n",
    "          \"Notch1\",\"Nr4a1\",\"Nr4a2\",\"Nr4a3\",\"NRP1\",\"Pdcd1\",\"Pou2f2\",\"Pparg\",\"Prdm1\",\"Prf1\",\"PTPN13\",\"Rel\",\"Relb\",\"Rora\",\n",
    "          \"Rorc\",\"Runx3\",\"S1pr1\",\"Sell\",\"SH2D1A\",\"SLAMF1\",\"Slamf6\",\"Smad2\",\"Smad3\",\"Smad4\",\"Smad7\",\"SPI1\",\"STAT1\",\"STAT3\",\n",
    "          \"STAT4\",\"STAT5A\",\"STAT6\",\"Tbx21\",\"Tcf7\",\"TGFB1\",\"Tgfb1\",\"Tgfbr1\",\"Tgfbr2\",\"Tigit\",\"TMEM121B\",\"Tnf\",\"TNFRSF18\",\n",
    "          \"TNFRSF4\",\"Tnfsf11\",\"VPS51\",\"VPS52\",\"VPS53\",\"VPS54\",\"Zeb2\",\"Tox\",\"Bhlhe40\",\"Bach2\",\"CD44\",\"ZBTB16\",\"KLRB1\")\n",
    "\n",
    "taggene=unique(toupper(taggene))\n",
    "marker_genes=readRDS(paste0(\"markerGenes_\",base_Rdata_name,\"_\",tag_celltype,\".rds\"))\n",
    "taggene=taggene[taggene %in% marker_genes$gene]\n",
    "\n",
    "tagExpressionNormed=ExpressionMat_by_gp[,which(colnames(ExpressionMat_by_gp) %in% taggene)]\n",
    "#missing_gene=paste0(paste0(taggene[!taggene %in% colnames(ExpressionMat_by_gp)][1:20],collapse = \", \"),\"\\n\",\n",
    "#                    paste0(taggene[!taggene %in% colnames(ExpressionMat_by_gp)][21:40],collapse = \", \"))\n",
    "missing_gene=\"\"\n",
    "message(missing_gene)\n",
    "\n",
    "rownames(tagExpressionNormed)=Tcell_assignment_expanded$subcelltype[order(as.numeric(Tcell_assignment_expanded$cluster))]\n",
    "\n",
    "options(repr.plot.width=30, repr.plot.height=8)\n",
    "\n",
    "pdf(paste0(\"clusterGeneExp_\",tag_celltype,\"_heatmap.pdf\"),width=25, height=8) #_magic\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "heatmap.2(scale(tagExpressionNormed),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(0.7,5),\n",
    "          main=paste0(\"cell assignment vs. gene\\nmissing:\",missing_gene),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "\n",
    "heatmap.2(as.matrix(tagExpressionNormed),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(0.7,5),\n",
    "          main=paste0(\"cell assignment vs. gene\\nmissing:\",missing_gene),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "dev.off()\n",
    "\n",
    "write.table(scale(tagExpressionNormed),file=paste0(\"GeneExpPerClusterScaled_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(tagExpressionNormed,file=paste0(\"GeneExpPerCluster_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### heatmap cell-DEG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "Group_type=unique(Tcellsubcelltype_assignment)\n",
    "Group_type=Group_type[gtools::mixedorder(Group_type)]\n",
    "\n",
    "marker_genes=readRDS(paste0(\"markerGenes_\",base_Rdata_name,\"_\",tag_celltype,\".rds\"))\n",
    "get5k=Subsample_by_group(Tcellsubcelltype_assignment,5000)\n",
    "toplot=ExpressionMat_magic[get5k,unique(marker_genes$gene)] #t(Exp_Seurat@data[unique(marker_genes$gene),get5k])\n",
    "\n",
    "tmp=as.numeric(gp_name_replacing(Tcellsubcelltype_assignment[get5k],Group_type,1:length(Group_type)))\n",
    "globall_ordering=fastcluster::hclust.vector(toplot,method='ward')\n",
    "globall_sorting_idx=match(1:length(get5k),globall_ordering$order)\n",
    "colOrder=order(tmp,Exp_Seurat@meta.data$disease_assignment[get5k],globall_sorting_idx)\n",
    "\n",
    "toplot=toplot[colOrder,]\n",
    "toplot=t(scale(toplot))\n",
    "toplot[toplot>1.5]=1.5\n",
    "toplot[toplot<(-1.5)]=-1.5\n",
    "\n",
    "ColforDisease=RColorBrewer::brewer.pal(3,\"Set1\")[2:1]\n",
    "ColforSubCelltype=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(Group_type))\n",
    "ColforPatient_ID=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(patient_type))\n",
    "ColSideColors=cbind(patient_ID=ColforPatient_ID[Exp_Seurat@meta.data$patient_assignment],\n",
    "                    IBD_status=ColforDisease[Exp_Seurat@meta.data$disease_assignment],\n",
    "                    subcelltype=ColforSubCelltype[match(Tcellsubcelltype_assignment,Group_type)])\n",
    "ColSideColors=ColSideColors[get5k[colOrder],]\n",
    "\n",
    "RowSideColors=matrix(\"white\",nrow(toplot),max(as.numeric(marker_genes$cluster)))\n",
    "colnames(RowSideColors)=Group_type\n",
    "for(i in 1:max(as.numeric(marker_genes$cluster))){\n",
    "    tagG=marker_genes$gene[marker_genes$cluster==i-1]\n",
    "    tagsubtype=Tcell_assignment_expanded$subcelltype[Tcell_assignment_expanded$cluster==i-1]\n",
    "    RowSideColors[rownames(toplot) %in% tagG,which(Group_type==tagsubtype)]=ColforSubCelltype[which(Group_type==tagsubtype)]\n",
    "}\n",
    "\n",
    "showgeneid=rownames(toplot)\n",
    "showgeneid[!showgeneid %in% GWAS_gene]=\"\"\n",
    "\n",
    "#ColSideColors can be a data.frame with each col indicate one color assignment\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "pdf(file=paste0(tag_celltype,\"_DEGbycluster_heatmap.pdf\"),width=15,height=10)\n",
    "heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,RowSideColors=RowSideColors,ColSideColors=ColSideColors,\n",
    "                   labRow=showgeneid,labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none')\n",
    "dev.off()\n",
    "\n",
    "png(file=paste0(tag_celltype,\"_DEGbycluster_heatmap.png\"),width=15,height=10,units = \"in\",res=150)\n",
    "heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,RowSideColors=RowSideColors,ColSideColors=ColSideColors,\n",
    "                   labRow=showgeneid,labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none')\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>''</li>\n",
       "\t<li>'FOS'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'TNFAIP3'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'IRF1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'ZFP36L1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'GPR183'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'PTGER4'</li>\n",
       "\t<li>'TAGAP'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'STAT4'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'PTPN22'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'MAP3K8'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'ATF4'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'LIME1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'GPX4'</li>\n",
       "\t<li>'RNASET2'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'LSP1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'TNFRSF18'</li>\n",
       "\t<li>'TNFRSF4'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'GPX1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'CREM'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'CALM3'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'STAT3'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'TNFRSF14'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'IFNG'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'HLA-DRB1'</li>\n",
       "\t<li>'HLA-C'</li>\n",
       "\t<li>'CTSW'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'FYN'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'LITAF'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'REL'</li>\n",
       "\t<li>'SOCS1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'ARPC2'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'NDFIP1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'CD48'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'IKZF1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'IL6ST'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'MAML2'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'RASGRP1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'LNPEP'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'NFKB1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'CD6'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'ITGAL'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'GPR65'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'TXK'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'UBE2L3'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'RASSF5'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'PHACTR2'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'CD5'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>'CTDSP1'</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "\t<li>''</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item ''\n",
       "\\item 'FOS'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'TNFAIP3'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'IRF1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'ZFP36L1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'GPR183'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'PTGER4'\n",
       "\\item 'TAGAP'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'STAT4'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'PTPN22'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'MAP3K8'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'ATF4'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'LIME1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'GPX4'\n",
       "\\item 'RNASET2'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'LSP1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'TNFRSF18'\n",
       "\\item 'TNFRSF4'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'GPX1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'CREM'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'CALM3'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'STAT3'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'TNFRSF14'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'IFNG'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'HLA-DRB1'\n",
       "\\item 'HLA-C'\n",
       "\\item 'CTSW'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'FYN'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'LITAF'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'REL'\n",
       "\\item 'SOCS1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'ARPC2'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'NDFIP1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'CD48'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'IKZF1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'IL6ST'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'MAML2'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'RASGRP1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'LNPEP'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'NFKB1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'CD6'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'ITGAL'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'GPR65'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'TXK'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'UBE2L3'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'RASSF5'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'PHACTR2'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'CD5'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item 'CTDSP1'\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\item ''\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. ''\n",
       "2. 'FOS'\n",
       "3. ''\n",
       "4. ''\n",
       "5. ''\n",
       "6. ''\n",
       "7. 'TNFAIP3'\n",
       "8. ''\n",
       "9. ''\n",
       "10. ''\n",
       "11. ''\n",
       "12. ''\n",
       "13. ''\n",
       "14. ''\n",
       "15. ''\n",
       "16. ''\n",
       "17. ''\n",
       "18. ''\n",
       "19. ''\n",
       "20. ''\n",
       "21. ''\n",
       "22. ''\n",
       "23. ''\n",
       "24. ''\n",
       "25. ''\n",
       "26. ''\n",
       "27. ''\n",
       "28. ''\n",
       "29. ''\n",
       "30. ''\n",
       "31. ''\n",
       "32. ''\n",
       "33. ''\n",
       "34. 'IRF1'\n",
       "35. ''\n",
       "36. ''\n",
       "37. ''\n",
       "38. ''\n",
       "39. ''\n",
       "40. ''\n",
       "41. ''\n",
       "42. ''\n",
       "43. ''\n",
       "44. ''\n",
       "45. ''\n",
       "46. ''\n",
       "47. 'ZFP36L1'\n",
       "48. ''\n",
       "49. ''\n",
       "50. 'GPR183'\n",
       "51. ''\n",
       "52. ''\n",
       "53. ''\n",
       "54. ''\n",
       "55. ''\n",
       "56. ''\n",
       "57. ''\n",
       "58. ''\n",
       "59. ''\n",
       "60. ''\n",
       "61. ''\n",
       "62. ''\n",
       "63. 'PTGER4'\n",
       "64. 'TAGAP'\n",
       "65. ''\n",
       "66. ''\n",
       "67. ''\n",
       "68. ''\n",
       "69. ''\n",
       "70. ''\n",
       "71. ''\n",
       "72. ''\n",
       "73. ''\n",
       "74. ''\n",
       "75. ''\n",
       "76. ''\n",
       "77. ''\n",
       "78. ''\n",
       "79. ''\n",
       "80. ''\n",
       "81. ''\n",
       "82. ''\n",
       "83. ''\n",
       "84. ''\n",
       "85. ''\n",
       "86. ''\n",
       "87. ''\n",
       "88. ''\n",
       "89. ''\n",
       "90. ''\n",
       "91. ''\n",
       "92. ''\n",
       "93. ''\n",
       "94. ''\n",
       "95. ''\n",
       "96. ''\n",
       "97. ''\n",
       "98. ''\n",
       "99. ''\n",
       "100. ''\n",
       "101. ''\n",
       "102. ''\n",
       "103. ''\n",
       "104. ''\n",
       "105. ''\n",
       "106. ''\n",
       "107. ''\n",
       "108. ''\n",
       "109. ''\n",
       "110. ''\n",
       "111. ''\n",
       "112. ''\n",
       "113. ''\n",
       "114. ''\n",
       "115. ''\n",
       "116. ''\n",
       "117. ''\n",
       "118. ''\n",
       "119. 'STAT4'\n",
       "120. ''\n",
       "121. ''\n",
       "122. ''\n",
       "123. ''\n",
       "124. ''\n",
       "125. ''\n",
       "126. ''\n",
       "127. ''\n",
       "128. ''\n",
       "129. ''\n",
       "130. ''\n",
       "131. ''\n",
       "132. ''\n",
       "133. ''\n",
       "134. ''\n",
       "135. ''\n",
       "136. 'PTPN22'\n",
       "137. ''\n",
       "138. ''\n",
       "139. ''\n",
       "140. ''\n",
       "141. ''\n",
       "142. ''\n",
       "143. ''\n",
       "144. ''\n",
       "145. ''\n",
       "146. ''\n",
       "147. ''\n",
       "148. 'MAP3K8'\n",
       "149. ''\n",
       "150. ''\n",
       "151. ''\n",
       "152. ''\n",
       "153. ''\n",
       "154. ''\n",
       "155. ''\n",
       "156. ''\n",
       "157. ''\n",
       "158. ''\n",
       "159. ''\n",
       "160. ''\n",
       "161. ''\n",
       "162. ''\n",
       "163. ''\n",
       "164. ''\n",
       "165. ''\n",
       "166. ''\n",
       "167. ''\n",
       "168. 'ATF4'\n",
       "169. ''\n",
       "170. ''\n",
       "171. ''\n",
       "172. ''\n",
       "173. ''\n",
       "174. ''\n",
       "175. ''\n",
       "176. ''\n",
       "177. ''\n",
       "178. ''\n",
       "179. ''\n",
       "180. ''\n",
       "181. ''\n",
       "182. ''\n",
       "183. ''\n",
       "184. ''\n",
       "185. ''\n",
       "186. ''\n",
       "187. ''\n",
       "188. ''\n",
       "189. ''\n",
       "190. ''\n",
       "191. ''\n",
       "192. ''\n",
       "193. ''\n",
       "194. ''\n",
       "195. ''\n",
       "196. ''\n",
       "197. ''\n",
       "198. ''\n",
       "199. ''\n",
       "200. ''\n",
       "201. ''\n",
       "202. ''\n",
       "203. ''\n",
       "204. ''\n",
       "205. ''\n",
       "206. ''\n",
       "207. ''\n",
       "208. ''\n",
       "209. ''\n",
       "210. ''\n",
       "211. ''\n",
       "212. ''\n",
       "213. ''\n",
       "214. ''\n",
       "215. ''\n",
       "216. ''\n",
       "217. ''\n",
       "218. ''\n",
       "219. ''\n",
       "220. ''\n",
       "221. ''\n",
       "222. ''\n",
       "223. ''\n",
       "224. ''\n",
       "225. ''\n",
       "226. ''\n",
       "227. ''\n",
       "228. ''\n",
       "229. ''\n",
       "230. ''\n",
       "231. ''\n",
       "232. ''\n",
       "233. ''\n",
       "234. ''\n",
       "235. ''\n",
       "236. ''\n",
       "237. ''\n",
       "238. ''\n",
       "239. ''\n",
       "240. ''\n",
       "241. ''\n",
       "242. ''\n",
       "243. ''\n",
       "244. ''\n",
       "245. ''\n",
       "246. ''\n",
       "247. ''\n",
       "248. ''\n",
       "249. ''\n",
       "250. ''\n",
       "251. ''\n",
       "252. ''\n",
       "253. ''\n",
       "254. ''\n",
       "255. ''\n",
       "256. ''\n",
       "257. ''\n",
       "258. ''\n",
       "259. ''\n",
       "260. ''\n",
       "261. ''\n",
       "262. ''\n",
       "263. ''\n",
       "264. ''\n",
       "265. ''\n",
       "266. ''\n",
       "267. ''\n",
       "268. ''\n",
       "269. ''\n",
       "270. ''\n",
       "271. ''\n",
       "272. ''\n",
       "273. ''\n",
       "274. ''\n",
       "275. ''\n",
       "276. 'LIME1'\n",
       "277. ''\n",
       "278. ''\n",
       "279. ''\n",
       "280. ''\n",
       "281. ''\n",
       "282. ''\n",
       "283. ''\n",
       "284. ''\n",
       "285. ''\n",
       "286. ''\n",
       "287. ''\n",
       "288. ''\n",
       "289. ''\n",
       "290. ''\n",
       "291. ''\n",
       "292. ''\n",
       "293. ''\n",
       "294. ''\n",
       "295. ''\n",
       "296. ''\n",
       "297. ''\n",
       "298. ''\n",
       "299. ''\n",
       "300. ''\n",
       "301. ''\n",
       "302. ''\n",
       "303. ''\n",
       "304. ''\n",
       "305. ''\n",
       "306. ''\n",
       "307. ''\n",
       "308. ''\n",
       "309. ''\n",
       "310. ''\n",
       "311. ''\n",
       "312. ''\n",
       "313. ''\n",
       "314. ''\n",
       "315. ''\n",
       "316. ''\n",
       "317. ''\n",
       "318. ''\n",
       "319. ''\n",
       "320. ''\n",
       "321. ''\n",
       "322. ''\n",
       "323. ''\n",
       "324. ''\n",
       "325. ''\n",
       "326. ''\n",
       "327. ''\n",
       "328. ''\n",
       "329. ''\n",
       "330. ''\n",
       "331. ''\n",
       "332. ''\n",
       "333. ''\n",
       "334. ''\n",
       "335. ''\n",
       "336. ''\n",
       "337. ''\n",
       "338. ''\n",
       "339. ''\n",
       "340. ''\n",
       "341. ''\n",
       "342. ''\n",
       "343. ''\n",
       "344. ''\n",
       "345. ''\n",
       "346. ''\n",
       "347. ''\n",
       "348. ''\n",
       "349. ''\n",
       "350. ''\n",
       "351. ''\n",
       "352. ''\n",
       "353. ''\n",
       "354. ''\n",
       "355. ''\n",
       "356. ''\n",
       "357. ''\n",
       "358. ''\n",
       "359. ''\n",
       "360. ''\n",
       "361. ''\n",
       "362. ''\n",
       "363. ''\n",
       "364. ''\n",
       "365. ''\n",
       "366. ''\n",
       "367. ''\n",
       "368. ''\n",
       "369. ''\n",
       "370. ''\n",
       "371. ''\n",
       "372. ''\n",
       "373. ''\n",
       "374. ''\n",
       "375. ''\n",
       "376. ''\n",
       "377. ''\n",
       "378. ''\n",
       "379. ''\n",
       "380. ''\n",
       "381. ''\n",
       "382. ''\n",
       "383. ''\n",
       "384. ''\n",
       "385. ''\n",
       "386. ''\n",
       "387. ''\n",
       "388. ''\n",
       "389. ''\n",
       "390. ''\n",
       "391. ''\n",
       "392. ''\n",
       "393. ''\n",
       "394. ''\n",
       "395. ''\n",
       "396. ''\n",
       "397. ''\n",
       "398. ''\n",
       "399. ''\n",
       "400. ''\n",
       "401. ''\n",
       "402. ''\n",
       "403. ''\n",
       "404. ''\n",
       "405. 'GPX4'\n",
       "406. 'RNASET2'\n",
       "407. ''\n",
       "408. ''\n",
       "409. ''\n",
       "410. ''\n",
       "411. ''\n",
       "412. ''\n",
       "413. ''\n",
       "414. ''\n",
       "415. ''\n",
       "416. ''\n",
       "417. ''\n",
       "418. 'LSP1'\n",
       "419. ''\n",
       "420. ''\n",
       "421. ''\n",
       "422. ''\n",
       "423. ''\n",
       "424. ''\n",
       "425. ''\n",
       "426. 'TNFRSF18'\n",
       "427. 'TNFRSF4'\n",
       "428. ''\n",
       "429. ''\n",
       "430. ''\n",
       "431. ''\n",
       "432. ''\n",
       "433. 'GPX1'\n",
       "434. ''\n",
       "435. ''\n",
       "436. ''\n",
       "437. ''\n",
       "438. ''\n",
       "439. ''\n",
       "440. ''\n",
       "441. ''\n",
       "442. ''\n",
       "443. ''\n",
       "444. ''\n",
       "445. ''\n",
       "446. 'CREM'\n",
       "447. ''\n",
       "448. ''\n",
       "449. ''\n",
       "450. ''\n",
       "451. ''\n",
       "452. ''\n",
       "453. ''\n",
       "454. ''\n",
       "455. ''\n",
       "456. ''\n",
       "457. ''\n",
       "458. ''\n",
       "459. ''\n",
       "460. ''\n",
       "461. ''\n",
       "462. ''\n",
       "463. ''\n",
       "464. ''\n",
       "465. ''\n",
       "466. ''\n",
       "467. ''\n",
       "468. ''\n",
       "469. ''\n",
       "470. ''\n",
       "471. ''\n",
       "472. ''\n",
       "473. ''\n",
       "474. ''\n",
       "475. ''\n",
       "476. ''\n",
       "477. ''\n",
       "478. ''\n",
       "479. ''\n",
       "480. ''\n",
       "481. ''\n",
       "482. ''\n",
       "483. ''\n",
       "484. 'CALM3'\n",
       "485. ''\n",
       "486. ''\n",
       "487. ''\n",
       "488. ''\n",
       "489. ''\n",
       "490. ''\n",
       "491. ''\n",
       "492. ''\n",
       "493. ''\n",
       "494. ''\n",
       "495. ''\n",
       "496. ''\n",
       "497. ''\n",
       "498. ''\n",
       "499. 'STAT3'\n",
       "500. ''\n",
       "501. ''\n",
       "502. ''\n",
       "503. ''\n",
       "504. ''\n",
       "505. ''\n",
       "506. ''\n",
       "507. ''\n",
       "508. ''\n",
       "509. ''\n",
       "510. ''\n",
       "511. ''\n",
       "512. ''\n",
       "513. ''\n",
       "514. ''\n",
       "515. ''\n",
       "516. ''\n",
       "517. ''\n",
       "518. ''\n",
       "519. ''\n",
       "520. ''\n",
       "521. ''\n",
       "522. ''\n",
       "523. ''\n",
       "524. ''\n",
       "525. ''\n",
       "526. ''\n",
       "527. ''\n",
       "528. ''\n",
       "529. ''\n",
       "530. ''\n",
       "531. ''\n",
       "532. ''\n",
       "533. ''\n",
       "534. ''\n",
       "535. ''\n",
       "536. ''\n",
       "537. ''\n",
       "538. ''\n",
       "539. ''\n",
       "540. ''\n",
       "541. ''\n",
       "542. ''\n",
       "543. ''\n",
       "544. ''\n",
       "545. ''\n",
       "546. ''\n",
       "547. ''\n",
       "548. ''\n",
       "549. ''\n",
       "550. ''\n",
       "551. ''\n",
       "552. ''\n",
       "553. ''\n",
       "554. ''\n",
       "555. ''\n",
       "556. ''\n",
       "557. ''\n",
       "558. ''\n",
       "559. ''\n",
       "560. ''\n",
       "561. ''\n",
       "562. ''\n",
       "563. ''\n",
       "564. ''\n",
       "565. ''\n",
       "566. ''\n",
       "567. ''\n",
       "568. ''\n",
       "569. ''\n",
       "570. ''\n",
       "571. ''\n",
       "572. ''\n",
       "573. ''\n",
       "574. ''\n",
       "575. ''\n",
       "576. ''\n",
       "577. ''\n",
       "578. ''\n",
       "579. ''\n",
       "580. ''\n",
       "581. ''\n",
       "582. ''\n",
       "583. ''\n",
       "584. ''\n",
       "585. ''\n",
       "586. ''\n",
       "587. ''\n",
       "588. ''\n",
       "589. ''\n",
       "590. ''\n",
       "591. ''\n",
       "592. ''\n",
       "593. ''\n",
       "594. 'TNFRSF14'\n",
       "595. ''\n",
       "596. ''\n",
       "597. ''\n",
       "598. ''\n",
       "599. ''\n",
       "600. ''\n",
       "601. ''\n",
       "602. ''\n",
       "603. ''\n",
       "604. ''\n",
       "605. ''\n",
       "606. ''\n",
       "607. ''\n",
       "608. ''\n",
       "609. ''\n",
       "610. ''\n",
       "611. ''\n",
       "612. ''\n",
       "613. ''\n",
       "614. ''\n",
       "615. 'IFNG'\n",
       "616. ''\n",
       "617. ''\n",
       "618. ''\n",
       "619. ''\n",
       "620. ''\n",
       "621. ''\n",
       "622. ''\n",
       "623. ''\n",
       "624. ''\n",
       "625. ''\n",
       "626. 'HLA-DRB1'\n",
       "627. 'HLA-C'\n",
       "628. 'CTSW'\n",
       "629. ''\n",
       "630. ''\n",
       "631. ''\n",
       "632. ''\n",
       "633. ''\n",
       "634. ''\n",
       "635. ''\n",
       "636. ''\n",
       "637. ''\n",
       "638. ''\n",
       "639. ''\n",
       "640. ''\n",
       "641. ''\n",
       "642. ''\n",
       "643. 'FYN'\n",
       "644. ''\n",
       "645. ''\n",
       "646. 'LITAF'\n",
       "647. ''\n",
       "648. ''\n",
       "649. ''\n",
       "650. ''\n",
       "651. ''\n",
       "652. ''\n",
       "653. ''\n",
       "654. ''\n",
       "655. ''\n",
       "656. ''\n",
       "657. ''\n",
       "658. ''\n",
       "659. ''\n",
       "660. 'REL'\n",
       "661. 'SOCS1'\n",
       "662. ''\n",
       "663. ''\n",
       "664. ''\n",
       "665. 'ARPC2'\n",
       "666. ''\n",
       "667. ''\n",
       "668. ''\n",
       "669. ''\n",
       "670. ''\n",
       "671. ''\n",
       "672. ''\n",
       "673. ''\n",
       "674. ''\n",
       "675. ''\n",
       "676. ''\n",
       "677. ''\n",
       "678. ''\n",
       "679. ''\n",
       "680. ''\n",
       "681. ''\n",
       "682. ''\n",
       "683. ''\n",
       "684. ''\n",
       "685. ''\n",
       "686. ''\n",
       "687. ''\n",
       "688. ''\n",
       "689. ''\n",
       "690. ''\n",
       "691. ''\n",
       "692. ''\n",
       "693. ''\n",
       "694. ''\n",
       "695. ''\n",
       "696. ''\n",
       "697. ''\n",
       "698. ''\n",
       "699. ''\n",
       "700. ''\n",
       "701. ''\n",
       "702. ''\n",
       "703. ''\n",
       "704. ''\n",
       "705. ''\n",
       "706. ''\n",
       "707. ''\n",
       "708. ''\n",
       "709. ''\n",
       "710. ''\n",
       "711. ''\n",
       "712. ''\n",
       "713. ''\n",
       "714. ''\n",
       "715. ''\n",
       "716. ''\n",
       "717. ''\n",
       "718. ''\n",
       "719. 'NDFIP1'\n",
       "720. ''\n",
       "721. ''\n",
       "722. ''\n",
       "723. ''\n",
       "724. ''\n",
       "725. ''\n",
       "726. ''\n",
       "727. ''\n",
       "728. ''\n",
       "729. 'CD48'\n",
       "730. ''\n",
       "731. ''\n",
       "732. ''\n",
       "733. ''\n",
       "734. ''\n",
       "735. ''\n",
       "736. ''\n",
       "737. ''\n",
       "738. ''\n",
       "739. ''\n",
       "740. ''\n",
       "741. ''\n",
       "742. ''\n",
       "743. ''\n",
       "744. ''\n",
       "745. ''\n",
       "746. ''\n",
       "747. ''\n",
       "748. ''\n",
       "749. ''\n",
       "750. ''\n",
       "751. ''\n",
       "752. ''\n",
       "753. ''\n",
       "754. ''\n",
       "755. ''\n",
       "756. ''\n",
       "757. ''\n",
       "758. ''\n",
       "759. ''\n",
       "760. ''\n",
       "761. ''\n",
       "762. ''\n",
       "763. ''\n",
       "764. 'IKZF1'\n",
       "765. ''\n",
       "766. ''\n",
       "767. ''\n",
       "768. ''\n",
       "769. ''\n",
       "770. ''\n",
       "771. ''\n",
       "772. ''\n",
       "773. ''\n",
       "774. ''\n",
       "775. ''\n",
       "776. ''\n",
       "777. ''\n",
       "778. ''\n",
       "779. ''\n",
       "780. ''\n",
       "781. ''\n",
       "782. ''\n",
       "783. ''\n",
       "784. ''\n",
       "785. ''\n",
       "786. ''\n",
       "787. ''\n",
       "788. ''\n",
       "789. ''\n",
       "790. ''\n",
       "791. 'IL6ST'\n",
       "792. ''\n",
       "793. ''\n",
       "794. ''\n",
       "795. ''\n",
       "796. ''\n",
       "797. ''\n",
       "798. ''\n",
       "799. ''\n",
       "800. ''\n",
       "801. ''\n",
       "802. ''\n",
       "803. ''\n",
       "804. ''\n",
       "805. ''\n",
       "806. ''\n",
       "807. ''\n",
       "808. ''\n",
       "809. ''\n",
       "810. ''\n",
       "811. ''\n",
       "812. ''\n",
       "813. ''\n",
       "814. ''\n",
       "815. ''\n",
       "816. ''\n",
       "817. ''\n",
       "818. ''\n",
       "819. ''\n",
       "820. ''\n",
       "821. ''\n",
       "822. ''\n",
       "823. ''\n",
       "824. ''\n",
       "825. ''\n",
       "826. 'MAML2'\n",
       "827. ''\n",
       "828. ''\n",
       "829. ''\n",
       "830. ''\n",
       "831. ''\n",
       "832. ''\n",
       "833. ''\n",
       "834. ''\n",
       "835. ''\n",
       "836. ''\n",
       "837. ''\n",
       "838. ''\n",
       "839. ''\n",
       "840. ''\n",
       "841. ''\n",
       "842. ''\n",
       "843. ''\n",
       "844. ''\n",
       "845. ''\n",
       "846. ''\n",
       "847. ''\n",
       "848. ''\n",
       "849. ''\n",
       "850. ''\n",
       "851. ''\n",
       "852. ''\n",
       "853. ''\n",
       "854. ''\n",
       "855. ''\n",
       "856. ''\n",
       "857. ''\n",
       "858. ''\n",
       "859. ''\n",
       "860. ''\n",
       "861. ''\n",
       "862. ''\n",
       "863. ''\n",
       "864. ''\n",
       "865. ''\n",
       "866. ''\n",
       "867. ''\n",
       "868. ''\n",
       "869. ''\n",
       "870. ''\n",
       "871. ''\n",
       "872. ''\n",
       "873. ''\n",
       "874. ''\n",
       "875. ''\n",
       "876. ''\n",
       "877. ''\n",
       "878. ''\n",
       "879. ''\n",
       "880. ''\n",
       "881. ''\n",
       "882. ''\n",
       "883. ''\n",
       "884. 'RASGRP1'\n",
       "885. ''\n",
       "886. ''\n",
       "887. ''\n",
       "888. ''\n",
       "889. ''\n",
       "890. ''\n",
       "891. ''\n",
       "892. ''\n",
       "893. ''\n",
       "894. ''\n",
       "895. ''\n",
       "896. ''\n",
       "897. ''\n",
       "898. ''\n",
       "899. ''\n",
       "900. ''\n",
       "901. 'LNPEP'\n",
       "902. ''\n",
       "903. ''\n",
       "904. ''\n",
       "905. ''\n",
       "906. ''\n",
       "907. ''\n",
       "908. ''\n",
       "909. ''\n",
       "910. ''\n",
       "911. ''\n",
       "912. ''\n",
       "913. ''\n",
       "914. ''\n",
       "915. ''\n",
       "916. ''\n",
       "917. ''\n",
       "918. ''\n",
       "919. ''\n",
       "920. ''\n",
       "921. ''\n",
       "922. ''\n",
       "923. ''\n",
       "924. ''\n",
       "925. ''\n",
       "926. ''\n",
       "927. ''\n",
       "928. ''\n",
       "929. ''\n",
       "930. ''\n",
       "931. ''\n",
       "932. ''\n",
       "933. ''\n",
       "934. ''\n",
       "935. ''\n",
       "936. ''\n",
       "937. ''\n",
       "938. ''\n",
       "939. ''\n",
       "940. ''\n",
       "941. ''\n",
       "942. ''\n",
       "943. ''\n",
       "944. ''\n",
       "945. 'NFKB1'\n",
       "946. ''\n",
       "947. ''\n",
       "948. ''\n",
       "949. ''\n",
       "950. ''\n",
       "951. ''\n",
       "952. ''\n",
       "953. ''\n",
       "954. ''\n",
       "955. ''\n",
       "956. ''\n",
       "957. ''\n",
       "958. ''\n",
       "959. ''\n",
       "960. ''\n",
       "961. ''\n",
       "962. ''\n",
       "963. ''\n",
       "964. ''\n",
       "965. ''\n",
       "966. ''\n",
       "967. ''\n",
       "968. ''\n",
       "969. ''\n",
       "970. ''\n",
       "971. ''\n",
       "972. ''\n",
       "973. ''\n",
       "974. ''\n",
       "975. ''\n",
       "976. ''\n",
       "977. ''\n",
       "978. ''\n",
       "979. ''\n",
       "980. ''\n",
       "981. ''\n",
       "982. ''\n",
       "983. ''\n",
       "984. ''\n",
       "985. ''\n",
       "986. ''\n",
       "987. ''\n",
       "988. ''\n",
       "989. ''\n",
       "990. ''\n",
       "991. ''\n",
       "992. ''\n",
       "993. ''\n",
       "994. ''\n",
       "995. ''\n",
       "996. ''\n",
       "997. ''\n",
       "998. ''\n",
       "999. ''\n",
       "1000. ''\n",
       "1001. ''\n",
       "1002. ''\n",
       "1003. ''\n",
       "1004. ''\n",
       "1005. ''\n",
       "1006. ''\n",
       "1007. ''\n",
       "1008. ''\n",
       "1009. ''\n",
       "1010. ''\n",
       "1011. ''\n",
       "1012. ''\n",
       "1013. ''\n",
       "1014. ''\n",
       "1015. ''\n",
       "1016. ''\n",
       "1017. 'CD6'\n",
       "1018. ''\n",
       "1019. ''\n",
       "1020. ''\n",
       "1021. ''\n",
       "1022. ''\n",
       "1023. ''\n",
       "1024. ''\n",
       "1025. ''\n",
       "1026. ''\n",
       "1027. ''\n",
       "1028. ''\n",
       "1029. ''\n",
       "1030. ''\n",
       "1031. ''\n",
       "1032. ''\n",
       "1033. ''\n",
       "1034. ''\n",
       "1035. ''\n",
       "1036. ''\n",
       "1037. ''\n",
       "1038. ''\n",
       "1039. ''\n",
       "1040. ''\n",
       "1041. ''\n",
       "1042. ''\n",
       "1043. ''\n",
       "1044. ''\n",
       "1045. ''\n",
       "1046. ''\n",
       "1047. ''\n",
       "1048. ''\n",
       "1049. ''\n",
       "1050. ''\n",
       "1051. ''\n",
       "1052. ''\n",
       "1053. ''\n",
       "1054. ''\n",
       "1055. ''\n",
       "1056. ''\n",
       "1057. ''\n",
       "1058. ''\n",
       "1059. ''\n",
       "1060. ''\n",
       "1061. ''\n",
       "1062. ''\n",
       "1063. ''\n",
       "1064. ''\n",
       "1065. ''\n",
       "1066. ''\n",
       "1067. ''\n",
       "1068. ''\n",
       "1069. ''\n",
       "1070. ''\n",
       "1071. ''\n",
       "1072. ''\n",
       "1073. ''\n",
       "1074. ''\n",
       "1075. ''\n",
       "1076. ''\n",
       "1077. ''\n",
       "1078. ''\n",
       "1079. ''\n",
       "1080. ''\n",
       "1081. ''\n",
       "1082. ''\n",
       "1083. 'ITGAL'\n",
       "1084. ''\n",
       "1085. ''\n",
       "1086. ''\n",
       "1087. ''\n",
       "1088. ''\n",
       "1089. ''\n",
       "1090. ''\n",
       "1091. ''\n",
       "1092. ''\n",
       "1093. ''\n",
       "1094. ''\n",
       "1095. ''\n",
       "1096. ''\n",
       "1097. ''\n",
       "1098. ''\n",
       "1099. ''\n",
       "1100. ''\n",
       "1101. ''\n",
       "1102. ''\n",
       "1103. ''\n",
       "1104. ''\n",
       "1105. ''\n",
       "1106. ''\n",
       "1107. ''\n",
       "1108. ''\n",
       "1109. ''\n",
       "1110. ''\n",
       "1111. ''\n",
       "1112. ''\n",
       "1113. ''\n",
       "1114. ''\n",
       "1115. ''\n",
       "1116. ''\n",
       "1117. ''\n",
       "1118. ''\n",
       "1119. ''\n",
       "1120. ''\n",
       "1121. ''\n",
       "1122. ''\n",
       "1123. ''\n",
       "1124. ''\n",
       "1125. ''\n",
       "1126. ''\n",
       "1127. ''\n",
       "1128. ''\n",
       "1129. ''\n",
       "1130. ''\n",
       "1131. ''\n",
       "1132. ''\n",
       "1133. ''\n",
       "1134. ''\n",
       "1135. ''\n",
       "1136. ''\n",
       "1137. ''\n",
       "1138. ''\n",
       "1139. ''\n",
       "1140. ''\n",
       "1141. ''\n",
       "1142. ''\n",
       "1143. ''\n",
       "1144. ''\n",
       "1145. ''\n",
       "1146. ''\n",
       "1147. ''\n",
       "1148. ''\n",
       "1149. ''\n",
       "1150. ''\n",
       "1151. ''\n",
       "1152. ''\n",
       "1153. ''\n",
       "1154. ''\n",
       "1155. ''\n",
       "1156. ''\n",
       "1157. ''\n",
       "1158. ''\n",
       "1159. ''\n",
       "1160. ''\n",
       "1161. ''\n",
       "1162. ''\n",
       "1163. ''\n",
       "1164. ''\n",
       "1165. ''\n",
       "1166. ''\n",
       "1167. ''\n",
       "1168. ''\n",
       "1169. ''\n",
       "1170. ''\n",
       "1171. ''\n",
       "1172. ''\n",
       "1173. ''\n",
       "1174. ''\n",
       "1175. ''\n",
       "1176. ''\n",
       "1177. ''\n",
       "1178. ''\n",
       "1179. ''\n",
       "1180. ''\n",
       "1181. ''\n",
       "1182. ''\n",
       "1183. ''\n",
       "1184. ''\n",
       "1185. ''\n",
       "1186. 'GPR65'\n",
       "1187. ''\n",
       "1188. ''\n",
       "1189. ''\n",
       "1190. ''\n",
       "1191. ''\n",
       "1192. ''\n",
       "1193. ''\n",
       "1194. ''\n",
       "1195. ''\n",
       "1196. ''\n",
       "1197. ''\n",
       "1198. ''\n",
       "1199. ''\n",
       "1200. ''\n",
       "1201. ''\n",
       "1202. 'TXK'\n",
       "1203. ''\n",
       "1204. ''\n",
       "1205. ''\n",
       "1206. ''\n",
       "1207. ''\n",
       "1208. ''\n",
       "1209. ''\n",
       "1210. ''\n",
       "1211. ''\n",
       "1212. ''\n",
       "1213. ''\n",
       "1214. ''\n",
       "1215. ''\n",
       "1216. ''\n",
       "1217. ''\n",
       "1218. ''\n",
       "1219. ''\n",
       "1220. ''\n",
       "1221. ''\n",
       "1222. ''\n",
       "1223. ''\n",
       "1224. ''\n",
       "1225. ''\n",
       "1226. ''\n",
       "1227. ''\n",
       "1228. ''\n",
       "1229. ''\n",
       "1230. ''\n",
       "1231. ''\n",
       "1232. ''\n",
       "1233. ''\n",
       "1234. ''\n",
       "1235. ''\n",
       "1236. ''\n",
       "1237. ''\n",
       "1238. ''\n",
       "1239. ''\n",
       "1240. ''\n",
       "1241. ''\n",
       "1242. ''\n",
       "1243. ''\n",
       "1244. ''\n",
       "1245. ''\n",
       "1246. ''\n",
       "1247. ''\n",
       "1248. ''\n",
       "1249. ''\n",
       "1250. 'UBE2L3'\n",
       "1251. ''\n",
       "1252. ''\n",
       "1253. ''\n",
       "1254. ''\n",
       "1255. ''\n",
       "1256. ''\n",
       "1257. ''\n",
       "1258. ''\n",
       "1259. ''\n",
       "1260. ''\n",
       "1261. ''\n",
       "1262. ''\n",
       "1263. ''\n",
       "1264. ''\n",
       "1265. ''\n",
       "1266. ''\n",
       "1267. ''\n",
       "1268. ''\n",
       "1269. ''\n",
       "1270. ''\n",
       "1271. ''\n",
       "1272. ''\n",
       "1273. ''\n",
       "1274. ''\n",
       "1275. 'RASSF5'\n",
       "1276. ''\n",
       "1277. ''\n",
       "1278. ''\n",
       "1279. ''\n",
       "1280. ''\n",
       "1281. ''\n",
       "1282. ''\n",
       "1283. 'PHACTR2'\n",
       "1284. ''\n",
       "1285. ''\n",
       "1286. ''\n",
       "1287. ''\n",
       "1288. ''\n",
       "1289. ''\n",
       "1290. ''\n",
       "1291. ''\n",
       "1292. ''\n",
       "1293. ''\n",
       "1294. ''\n",
       "1295. ''\n",
       "1296. ''\n",
       "1297. ''\n",
       "1298. ''\n",
       "1299. ''\n",
       "1300. ''\n",
       "1301. ''\n",
       "1302. ''\n",
       "1303. ''\n",
       "1304. ''\n",
       "1305. ''\n",
       "1306. ''\n",
       "1307. ''\n",
       "1308. ''\n",
       "1309. ''\n",
       "1310. ''\n",
       "1311. ''\n",
       "1312. ''\n",
       "1313. ''\n",
       "1314. ''\n",
       "1315. ''\n",
       "1316. ''\n",
       "1317. ''\n",
       "1318. ''\n",
       "1319. ''\n",
       "1320. ''\n",
       "1321. ''\n",
       "1322. ''\n",
       "1323. ''\n",
       "1324. ''\n",
       "1325. ''\n",
       "1326. ''\n",
       "1327. ''\n",
       "1328. ''\n",
       "1329. ''\n",
       "1330. ''\n",
       "1331. ''\n",
       "1332. ''\n",
       "1333. ''\n",
       "1334. ''\n",
       "1335. ''\n",
       "1336. ''\n",
       "1337. ''\n",
       "1338. ''\n",
       "1339. ''\n",
       "1340. ''\n",
       "1341. ''\n",
       "1342. ''\n",
       "1343. ''\n",
       "1344. ''\n",
       "1345. ''\n",
       "1346. ''\n",
       "1347. ''\n",
       "1348. ''\n",
       "1349. ''\n",
       "1350. ''\n",
       "1351. ''\n",
       "1352. ''\n",
       "1353. 'CD5'\n",
       "1354. ''\n",
       "1355. ''\n",
       "1356. ''\n",
       "1357. ''\n",
       "1358. 'CTDSP1'\n",
       "1359. ''\n",
       "1360. ''\n",
       "1361. ''\n",
       "1362. ''\n",
       "1363. ''\n",
       "1364. ''\n",
       "1365. ''\n",
       "1366. ''\n",
       "1367. ''\n",
       "1368. ''\n",
       "1369. ''\n",
       "1370. ''\n",
       "1371. ''\n",
       "1372. ''\n",
       "1373. ''\n",
       "1374. ''\n",
       "1375. ''\n",
       "1376. ''\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "   [1] \"\"         \"FOS\"      \"\"         \"\"         \"\"         \"\"        \n",
       "   [7] \"TNFAIP3\"  \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [13] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [19] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [25] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [31] \"\"         \"\"         \"\"         \"IRF1\"     \"\"         \"\"        \n",
       "  [37] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [43] \"\"         \"\"         \"\"         \"\"         \"ZFP36L1\"  \"\"        \n",
       "  [49] \"\"         \"GPR183\"   \"\"         \"\"         \"\"         \"\"        \n",
       "  [55] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [61] \"\"         \"\"         \"PTGER4\"   \"TAGAP\"    \"\"         \"\"        \n",
       "  [67] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [73] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [79] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [85] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [91] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "  [97] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [103] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [109] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [115] \"\"         \"\"         \"\"         \"\"         \"STAT4\"    \"\"        \n",
       " [121] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [127] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [133] \"\"         \"\"         \"\"         \"PTPN22\"   \"\"         \"\"        \n",
       " [139] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [145] \"\"         \"\"         \"\"         \"MAP3K8\"   \"\"         \"\"        \n",
       " [151] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [157] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [163] \"\"         \"\"         \"\"         \"\"         \"\"         \"ATF4\"    \n",
       " [169] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [175] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [181] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [187] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [193] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [199] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [205] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [211] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [217] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [223] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [229] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [235] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [241] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [247] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [253] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [259] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [265] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [271] \"\"         \"\"         \"\"         \"\"         \"\"         \"LIME1\"   \n",
       " [277] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [283] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [289] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [295] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [301] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [307] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [313] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [319] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [325] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [331] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [337] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [343] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [349] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [355] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [361] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [367] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [373] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [379] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [385] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [391] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [397] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [403] \"\"         \"\"         \"GPX4\"     \"RNASET2\"  \"\"         \"\"        \n",
       " [409] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [415] \"\"         \"\"         \"\"         \"LSP1\"     \"\"         \"\"        \n",
       " [421] \"\"         \"\"         \"\"         \"\"         \"\"         \"TNFRSF18\"\n",
       " [427] \"TNFRSF4\"  \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [433] \"GPX1\"     \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [439] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [445] \"\"         \"CREM\"     \"\"         \"\"         \"\"         \"\"        \n",
       " [451] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [457] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [463] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [469] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [475] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [481] \"\"         \"\"         \"\"         \"CALM3\"    \"\"         \"\"        \n",
       " [487] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [493] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [499] \"STAT3\"    \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [505] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [511] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [517] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [523] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [529] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [535] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [541] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [547] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [553] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [559] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [565] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [571] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [577] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [583] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [589] \"\"         \"\"         \"\"         \"\"         \"\"         \"TNFRSF14\"\n",
       " [595] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [601] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [607] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [613] \"\"         \"\"         \"IFNG\"     \"\"         \"\"         \"\"        \n",
       " [619] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [625] \"\"         \"HLA-DRB1\" \"HLA-C\"    \"CTSW\"     \"\"         \"\"        \n",
       " [631] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [637] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [643] \"FYN\"      \"\"         \"\"         \"LITAF\"    \"\"         \"\"        \n",
       " [649] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [655] \"\"         \"\"         \"\"         \"\"         \"\"         \"REL\"     \n",
       " [661] \"SOCS1\"    \"\"         \"\"         \"\"         \"ARPC2\"    \"\"        \n",
       " [667] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [673] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [679] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [685] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [691] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [697] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [703] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [709] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [715] \"\"         \"\"         \"\"         \"\"         \"NDFIP1\"   \"\"        \n",
       " [721] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [727] \"\"         \"\"         \"CD48\"     \"\"         \"\"         \"\"        \n",
       " [733] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [739] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [745] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [751] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [757] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [763] \"\"         \"IKZF1\"    \"\"         \"\"         \"\"         \"\"        \n",
       " [769] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [775] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [781] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [787] \"\"         \"\"         \"\"         \"\"         \"IL6ST\"    \"\"        \n",
       " [793] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [799] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [805] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [811] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [817] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [823] \"\"         \"\"         \"\"         \"MAML2\"    \"\"         \"\"        \n",
       " [829] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [835] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [841] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [847] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [853] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [859] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [865] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [871] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [877] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [883] \"\"         \"RASGRP1\"  \"\"         \"\"         \"\"         \"\"        \n",
       " [889] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [895] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [901] \"LNPEP\"    \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [907] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [913] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [919] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [925] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [931] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [937] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [943] \"\"         \"\"         \"NFKB1\"    \"\"         \"\"         \"\"        \n",
       " [949] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [955] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [961] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [967] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [973] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [979] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [985] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [991] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       " [997] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1003] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1009] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1015] \"\"         \"\"         \"CD6\"      \"\"         \"\"         \"\"        \n",
       "[1021] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1027] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1033] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1039] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1045] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1051] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1057] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1063] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1069] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1075] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1081] \"\"         \"\"         \"ITGAL\"    \"\"         \"\"         \"\"        \n",
       "[1087] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1093] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1099] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1105] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1111] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1117] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1123] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1129] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1135] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1141] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1147] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1153] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1159] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1165] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1171] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1177] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1183] \"\"         \"\"         \"\"         \"GPR65\"    \"\"         \"\"        \n",
       "[1189] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1195] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1201] \"\"         \"TXK\"      \"\"         \"\"         \"\"         \"\"        \n",
       "[1207] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1213] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1219] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1225] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1231] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1237] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1243] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1249] \"\"         \"UBE2L3\"   \"\"         \"\"         \"\"         \"\"        \n",
       "[1255] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1261] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1267] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1273] \"\"         \"\"         \"RASSF5\"   \"\"         \"\"         \"\"        \n",
       "[1279] \"\"         \"\"         \"\"         \"\"         \"PHACTR2\"  \"\"        \n",
       "[1285] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1291] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1297] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1303] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1309] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1315] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1321] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1327] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1333] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1339] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1345] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1351] \"\"         \"\"         \"CD5\"      \"\"         \"\"         \"\"        \n",
       "[1357] \"\"         \"CTDSP1\"   \"\"         \"\"         \"\"         \"\"        \n",
       "[1363] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1369] \"\"         \"\"         \"\"         \"\"         \"\"         \"\"        \n",
       "[1375] \"\"         \"\"        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "showgeneid"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### VDJ vs RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0(\"VDJassignment_vs_RNAassignment_\",tag_celltype,\"_heatmap.pdf\"),width=6,height=9)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "tmp=table(Tcellsubcelltype_assignment,Exp_Seurat@meta.data$cell_assignment)\n",
    "\n",
    "heatmap.2(t(apply(tmp,1,function(x){x/sum(x)})),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=F,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"RNA-based cell type assignment\\n vs. VDJ based\"),key.title=\"\",key.xlab=\"\",margins=c(15,10),cellnote=tmp)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### DEG within T cell groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tag_cluster=Tcell_assignment_expanded$cluster[Tcell_assignment_expanded$celltype %in% \"CD8\"] #c(\"CD8_TEM_effector\",\"CD8\",\"CD8_TCM\")\n",
    "cluster2rm=setdiff(celltype_assignment_expanded$cluster,tag_cluster)\n",
    "Exp_Seurat_T = filter_ident(Exp_Seurat,cluster2rm)\n",
    "marker_genes_CD8subcluster=FindAllMarkers(object = Exp_Seurat_T, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)\n",
    "marker_genes_CD8subcluster$cluster=gp_name_replacing(as.character(marker_genes_CD8subcluster$cluster),Tcell_assignment_expanded$cluster,Tcell_assignment_expanded$subcelltype)\n",
    "write_DEG(marker_genes_CD8subcluster,\"CD8\")\n",
    "\n",
    "tag_cluster=Tcell_assignment_expanded$cluster[Tcell_assignment_expanded$celltype %in% c(\"CD4\")]\n",
    "cluster2rm=setdiff(celltype_assignment_expanded$cluster,tag_cluster)\n",
    "Exp_Seurat_T = filter_ident(Exp_Seurat,cluster2rm)\n",
    "marker_genes_CD4subcluster=FindAllMarkers(object = Exp_Seurat_T, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)\n",
    "marker_genes_CD4subcluster$cluster=gp_name_replacing(as.character(marker_genes_CD4subcluster$cluster),Tcell_assignment_expanded$cluster,Tcell_assignment_expanded$subcelltype)\n",
    "write_DEG(marker_genes_CD4subcluster,\"CD4\")\n",
    "\n",
    "tag_cluster=Tcell_assignment_expanded$cluster[Tcell_assignment_expanded$celltype %in% c(\"NKT\")]\n",
    "cluster2rm=setdiff(celltype_assignment_expanded$cluster,tag_cluster)\n",
    "Exp_Seurat_T = filter_ident(Exp_Seurat,cluster2rm)\n",
    "marker_genes_nktsubcluster=FindAllMarkers(object = Exp_Seurat_T, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)\n",
    "marker_genes_nktsubcluster$cluster=gp_name_replacing(as.character(marker_genes_nktsubcluster$cluster),Tcell_assignment_expanded$cluster,Tcell_assignment_expanded$subcelltype)\n",
    "write_DEG(marker_genes_nktsubcluster,\"NKT\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in plotByGroup(Exp_Seurat, Group_assignment = expansion_info_summ$expand_lvl[match(colnames(Exp_Seurat@data), :\n",
      "“NAs in Group_assignment”"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "position_index_matrix=readRDS(\"T_cell_position_index_matrix.rds\")\n",
    "expansion_info_summ=data.table(position_index_matrix[!is.na(position_index_matrix$celltype),c(\"barcode\",\"expand_lvl\")])\n",
    "expansion_info_summ=expansion_info_summ[,.(expand_lvl=max(expand_lvl)),by=barcode]\n",
    "expansion_info_summ$expand_lvl[expansion_info_summ$expand_lvl>10]=\"10+\"\n",
    "\n",
    "pdf(paste0(\"TSNE_mito_\",tag_celltype,\".pdf\"),width=18,height=15)\n",
    "par(mfrow=c(3,3))\n",
    "plotByGroup(Exp_Seurat,Group_assignment=sample_list[Exp_Seurat@meta.data$sample_gp],\n",
    "            Group_type=sample_list,main=\"Batchs\") #batchs\n",
    "plotByGroup(Exp_Seurat,Group_assignment=tissue_type[Exp_Seurat@meta.data$tissue_assignment],\n",
    "            Group_type=tissue_type,main=\"tissue type\") #tissue\n",
    "plotByGroup(Exp_Seurat,Group_assignment=patient_type[Exp_Seurat@meta.data$patient_assignment],\n",
    "            Group_type=patient_type,main=\"patient id\") #patient id\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@meta.data$cell_assignment,\n",
    "            Group_type=c('failed BCR/TCR','B cell','T cell','dual-label','unknown'),main=\"B/T classification based on VDJ\",backGroundGroup=\"unknown\") #BT\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@ident,\n",
    "            Group_type=levels(Exp_Seurat@ident),main=\"cluster id\",Addtext=T) \n",
    "#plotByGroup(Exp_Seurat,Group_assignment=subcelltype_assignment,\n",
    "#            Group_type=celltype_assignment_expanded$subcelltype,main=\"rough cell type\",Addtext=T) \n",
    "plotByGroup(Exp_Seurat,Group_assignment=Tcellsubcelltype_assignment,\n",
    "            main=\"T cell subtype\",Addtext=T)\n",
    "plotByGroup(Exp_Seurat,Group_assignment=c(\"healthy\",\"diseased\")[Exp_Seurat@meta.data$disease_assignment],\n",
    "            Group_type=c(\"diseased\",\"healthy\"),main=\"disease condition\") #patient id\n",
    "plotByGroup(Exp_Seurat,Group_assignment=expansion_info_summ$expand_lvl[match(colnames(Exp_Seurat@data),expansion_info_summ$barcode)],\n",
    "            Group_type=c(1:9,\"10+\"),main=\"expansion level\",tagcol=colorRampPalette(rev(RColorBrewer::brewer.pal(7,\"RdBu\")))(10)) \n",
    "#plotByScore(Exp_Seurat,Score_assignment=Exp_Seurat@meta.data$S.Score,main=\"\") \n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "clusterVsPatient=table(Tcellsubcelltype_assignment,Exp_Seurat@meta.data$sample_gp)\n",
    "colnames(clusterVsPatient)=sample_list[as.numeric(colnames(clusterVsPatient))]\n",
    "clusterVsPatient_colPercent=round(apply(clusterVsPatient,2,function(x){x/sum(x)}),2)\n",
    "clusterVsPatient_rowPercent=round(t(apply(clusterVsPatient,1,function(x){x/sum(x)})),2)\n",
    "write.table(clusterVsPatient,file=paste0(\"clusterVsPatient_table_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(clusterVsPatient_rowPercent,file=paste0(\"clusterVsPatient_ClusterPercentage_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(clusterVsPatient_colPercent,file=paste0(\"clusterVsPatient_PatientPercentage_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "clusterVsPatient_rowPercent[clusterVsPatient_rowPercent>0.2]=0.2\n",
    "clusterVsPatient_colPercent[clusterVsPatient_colPercent>0.5]=0.5\n",
    "\n",
    "pdf(paste0(\"clusterSample_\",tag_celltype,\"_heatmap.pdf\"),width=15, height=8)\n",
    "heatmap.2(clusterVsPatient_colPercent,col=colorRampPalette(col2),\n",
    "          lhei=c(1,5),lwid=c(0.1,5),margins = c(10,10),trace=\"none\",density.info=\"none\",dendrogram='col',Colv=T,Rowv=T,\n",
    "          notecol=\"black\",key=F,cellnote=clusterVsPatient)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### per gene plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "require(feather)\n",
    "#MAGIC\n",
    "forMagic=\"For_magic.feather\"\n",
    "afterMagic=\"Magic_out.feather\"\n",
    "\n",
    "temp_Exp=as.data.frame(t(Exp_Seurat@data))\n",
    "#magic\n",
    "write_feather(temp_Exp, forMagic)\n",
    "system(paste0(\"python /home/ahe/tools/Lightbulb/src/Lightbulb_MagicWrapper.py --matx \",forMagic,\" --out \",afterMagic))\n",
    "ExpressionMat_magic = data.matrix(read_feather(afterMagic))\n",
    "colnames(ExpressionMat_magic)=gsub(\"MAGIC \", \"\", colnames(ExpressionMat_magic))\n",
    "\n",
    "saveRDS(ExpressionMat_magic,file=paste0(tag_celltype,\"_magic.rds\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CXCL13, IL10, IL4RA, KLRE1, NEURL3, TRAV, TRBV, ZFP683\n",
      "B3GAT1, CCR3, CCR5, CCR8, PTGDR2, EBI3, CSF2, HAVCR1, IL10, IL12A, IL12B, IL13, IL17A, IL17B, IL17F, IL17RE, IL1R1, IL1RL1, IL21, IL22, IL25, IL26, IL27, IL3, IL31, EBI3, IL4, IL5, IL6, IL9, IL9R, KLRE1, LTK, NRP1, PPARG, RORC, TMEM121B, TNFSF11\n",
      "CCR5\n",
      "CCR3, PPARG, IL5, IL1RL1, IL13, IL4, IL9R, IL3, IL5, IL6, IL10, IL13, IL25, IL31, CRTH2, CCR8, HAVCR1\n",
      "IL21, IL17A, IL1R1, IL17RE, RORC, IL26, IL22, IL17F\n",
      "TMEM121B, TNFSF11, TGFB, IL35, IL10, NRP1\n",
      "IL13, IL22\n",
      "IL6, IL10, IL21, B3GAT1\n",
      "IL9, IL10\n"
     ]
    }
   ],
   "source": [
    "ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "taggene=list()\n",
    "taggene[[1]]=c(\"Ahr\",\"Bach2\",\"Batf\",\"Bcl2\",\"Bcl6\",\"Bhlhe40\",\"Ccr7\",\"Cd101\",\"Cd244\",\"Cd28\",\"Cd3e\",\"Cd3g\",\"CD247\",\"CD4\",\"SPN\",\n",
    "               \"Cd69\",\"Cd8a\",\"Cd8b\",\"Crtam\",\"Ctla4\",\"Cx3cr1\",\"Cxcl13\",\"Cxcr3\",\"Cxcr5\",\"Cxcr6\",\"Dusp2\",\"Dusp5\",\"Dusp6\",\"Eomes\",\n",
    "               \"Ezh2\",\"Faslg\",\"Foxo1\",\"Gzma\",\"Gzmb\",\"Havcr2\",\"Icos\",\"Id2\",\"Id3\",\"Ifng\",\"Ifngr1\",\"Il10\",\"Il2\",\"Il21r\",\"Il2ra\",\n",
    "               \"Il2rb\",\"Il4ra\",\"Il7r\",\"Irf4\",\"Itga1\",\"Itga4\",\"Itgae\",\"Itgb7\",\"Itk\",\"Junb\",\"Jund\",\"Klf2\",\"Klf3\",\"Klf6\",\"Klre1\",\n",
    "               \"Klrg1\",\"Lag3\",\"Lat\",\"Lef1\",\"Litaf\",\"Malat1\",\"Myc\",\"Neurl3\",\"Notch1\",\"Nr4a1\",\"Nr4a2\",\"Nr4a3\",\"P2rx7\",\"Pdcd1\",\n",
    "               \"Pdcd4\",\"Pnrc1\",\"Prdm1\",\"Rbm39\",\"Rgs1\",\"Runx3\",\"S1pr1\",\"S1pr4\",\"S1pr5\",\"Sell\",\"Slamf6\",\"Socs3\",\"Stk17b\",\"Tbx21\",\n",
    "               \"Tcf7\",\"TRAV\",\"TRBV\",\"Tgfbr1\",\"Tgfbr2\",\"Tigit\",\"Tnf\",\"Tnfaip3\",\"Tox\",\"Ubc\",\"Vps37b\",\"Xist\",\"Zeb2\",\"Zfp683\")\n",
    "taggene[[2]]=c(\"ABCB1\",\"AHR\",\"Areg\",\"Asb2\",\"B3GAT1\",\"BATF\",\"Bcl6\",\"Ccl20\",\"CCL5\",\"Ccr1\",\"CCR10\",\"CCR3\",\"CCR4\",\"CCR5\",\"CCR6\",\n",
    "               \"CCR7\",\"CCR8\",\"CD27\",\"HAVCR2\",\"CD38\",\"ENTPD1\",\"Cd3e\",\"Cd3g\",\"CD247\",\"CD4\",\"CD40LG\",\"Cd69\",\"NT5E\",\"CD84\",\"Cd8a\",\n",
    "               \"Cd8b\",\"PTGDR2\",\"CTLA4\",\"CTSH\",\"Cx3cr1\",\"CXCR3\",\"CXCR5\",\"Cxcr6\",\"EBI3\",\"Eomes\",\"Ezh2\",\"FAS\",\"FaslG\",\"Foxo1\",\"Foxp3\",\n",
    "               \"Gata3\",\"CSF2\",\"Gzmb\",\"GZMK\",\"HAVCR1\",\"Havcr2\",\"HIF1A\",\"ICOS\",\"Id2\",\"Id3\",\"Ifng\",\"IFNGR1\",\"IKZF1\",\"Ikzf2\",\"IL10\",\n",
    "               \"IL12A\",\"Il12B\",\"Il12rb2\",\"Il13\",\"IL15\",\"Il17A\",\"Il17B\",\"IL17F\",\"IL17RA\",\"IL17re\",\"Il18\",\"IL18R1\",\"Il18RAP\",\n",
    "               \"IL1r1\",\"IL1Rl1\",\"IL2\",\"Il21\",\"Il21r\",\"IL22\",\"IL23A\",\"IL23R\",\"IL25\",\"IL26\",\"IL27\",\"IL27RA\",\"Il2RA\",\"IL3\",\"IL31\",\n",
    "               \"EBI3\",\"IL4\",\"IL4i1\",\"Il5\",\"IL6\",\"IL6R\",\"IL7R\",\"IL9\",\"Il9r\",\"IRF4\",\"IRF8\",\"Itga1\",\"Itga4\",\"Itgae\",\"Itgb7\",\"Klf2\",\n",
    "               \"Klf3\",\"Klf6\",\"KLRB1\",\"Klre1\",\"Klrg1\",\"Lag3\",\"LAMP1\",\"Lat\",\"Lef1\",\"Litaf\",\"LTA\",\"LTBR\",\"LTK\",\"MAF\",\"Malat1\",\"Myc\",\n",
    "               \"Notch1\",\"Nr4a1\",\"Nr4a2\",\"Nr4a3\",\"NRP1\",\"Pdcd1\",\"Pou2f2\",\"Pparg\",\"Prdm1\",\"Prf1\",\"PTPN13\",\"Rel\",\"Relb\",\"Rora\",\n",
    "               \"Rorc\",\"Runx3\",\"S1pr1\",\"Sell\",\"SH2D1A\",\"SLAMF1\",\"Slamf6\",\"Smad2\",\"Smad3\",\"Smad4\",\"Smad7\",\"SPI1\",\"STAT1\",\"STAT3\",\n",
    "               \"STAT4\",\"STAT5A\",\"STAT6\",\"Tbx21\",\"Tcf7\",\"TGFB1\",\"Tgfb1\",\"Tgfbr1\",\"Tgfbr2\",\"Tigit\",\"TMEM121B\",\"Tnf\",\"TNFRSF18\",\n",
    "               \"TNFRSF4\",\"Tnfsf11\",\"VPS51\",\"VPS52\",\"VPS53\",\"VPS54\",\"Zeb2\",\"Tox\",\"Bhlhe40\",\"Bach2\")\n",
    "taggene[[3]]=c(\"CCR5\",\"CXCR3\",\"Tbx21\",\"Eomes\",\"Il18R1\",\"FaslG\",\"Il12rb2\",\"Ifng\",\"Il18RAP\",\"STAT1\",\"STAT4\",\"HAVCR2\",\"LTA\",\"IL2\",\n",
    "               \"LTBR\",\"IFNGR1\",\"IRF8\")\n",
    "taggene[[4]]=c(\"CCR4\",\"CCR3\",\"Asb2\",\"Pparg\",\"Il5\",\"Ccr1\",\"Il1rl1\",\"Areg\",\"Il13\",\"IL4\",\"Il9r\",\"Gata3\",\"IL3\",\"IL5\",\"IL6\",\"IL10\",\n",
    "               \"IL13\",\"IL25\",\"IL31\",\"STAT6\",\"MAF\",\"CRTH2\",\"CCR8\",\"HAVCR1\",\"IFNGR1\",\"Prdm1\",\"IKZF1\",\"IRF4\")\n",
    "taggene[[5]]=c(\"Il21\",\"Il17A\",\"IL1r1\",\"IL17re\",\"Rorc\",\"Rora\",\"IL26\",\"IL22\",\"IL17F\",\"STAT3\",\"CCR4\",\"CCR6\",\"KLRB1\",\"IL23R\",\"BATF\",\n",
    "               \"MAF\",\"HIF1A\",\"IRF4\",\"IL17RA\")\n",
    "taggene[[6]]=c(\"Ccr6\",\"Foxp3\",\"TMEM121B\",\"Nr4a3\",\"Nr4a1\",\"Rel\",\"Relb\",\"Il2RA\",\"Pou2f2\",\"Tnfsf11\",\"Ikzf2\",\"KLRB1\",\"Sell\",\"IL7R\",\n",
    "               \"CD40LG\",\"Ccl20\",\"TGFb\",\"IL35\",\"IL10\",\"ENTPD1\",\"NT5E\",\"Itgae\",\"CTLA4\",\"TNFRSF18\",\"Smad4\",\"STAT5A\",\"VPS54\",\"VPS52\",\"VPS53\",\"VPS51\",\"Smad7\",\"Smad2\",\"Smad3\",\"NRP1\",\"Gzmb\",\"Prf1\",\"Tgfb1\",\"TNFRSF4\",\"Pdcd1\")\n",
    "taggene[[7]]=c(\"Il13\",\"IL22\",\"CCR4\",\"CCR6\",\"CCR10\",\"AHR\",\"STAT3\")\n",
    "taggene[[8]]=c(\"IL6\",\"Il10\",\"IL21\",\"CD84\",\"IL6R\",\"SLAMF1\",\"CD40LG\",\"CXCR5\",\"TNFRSF4\",\"ICOS\",\"PDCD1\",\"Bcl6\",\"SH2D1A\",\"FAS\",\"B3GAT1\",\"CCR7\")\n",
    "taggene[[9]]=c(\"GATA3\",\"STAT6\",\"IRF4\",\"IL9\",\"IL10\",\"BATF\",\"IRF4\",\"SPI1\")\n",
    "names(taggene)=c(\"CD8\",\"CD4\",\"Th1\",\"Th2\",\"Th17\",\"Treg\",\"Th22\",\"Tfh\",\"Th9\")\n",
    "missing_gene=list()\n",
    "tagExpressionNormed=list()\n",
    "for(i in 1:length(taggene)){\n",
    "    taggene[[i]]=toupper(taggene[[i]])\n",
    "    tagExpressionNormed[[i]]=ExpressionMat_magic[,which(colnames(ExpressionMat_magic) %in% taggene[[i]])]\n",
    "    tagExpressionNormed[[i]]=tagExpressionNormed[[i]][,order(colnames(tagExpressionNormed[[i]]))]\n",
    "    missing_gene[[i]]=paste0(taggene[[i]][!taggene[[i]] %in% colnames(tagExpressionNormed[[i]])],collapse = \", \")\n",
    "    message(missing_gene[[i]])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rbPal <- colorRampPalette(c('gray','#2b83ba','#abdda4','#ffffbf','#fdae61','#d7191c'))\n",
    "selected=sample(1:nrow(tagExpressionNormed[[1]]),2000)\n",
    "for(x in 1:length(taggene)){\n",
    "    pdf(paste0(tag_celltype,\"_\",names(taggene)[x],\"_per gene mito TSNE_magic.pdf\"),width=22,height=15)\n",
    "    par(mfrow=c(4,6))\n",
    "    for(i in 1:ncol(tagExpressionNormed[[x]])){\n",
    "        currentid=colnames(tagExpressionNormed[[x]])[i]\n",
    "        #calculate breaks\n",
    "        lowerbound=sort(tagExpressionNormed[[x]][selected,i])[round(length(selected)/33)]\n",
    "        upperbound=sort(tagExpressionNormed[[x]][selected,i])[round(length(selected)/33*32)]\n",
    "        if(upperbound!=lowerbound){\n",
    "            breaks=c(min(tagExpressionNormed[[x]][selected,i])-1,seq(lowerbound, upperbound, length.out=25),max(tagExpressionNormed[[x]][selected,i])+1)\n",
    "            cols=rbPal(25)[as.numeric(cut(tagExpressionNormed[[x]][selected,i],breaks = breaks, include.lowest=TRUE))]\n",
    "        }else{\n",
    "            breaks=c(min(tagExpressionNormed[[x]][selected,i])-1,max(tagExpressionNormed[[x]][selected,i])/2,max(tagExpressionNormed[[x]][selected,i]))\n",
    "            cols=rbPal(2)[as.numeric(cut(tagExpressionNormed[[x]][selected,i],breaks = breaks, include.lowest=TRUE))]\n",
    "        }\n",
    "        plot(Exp_Seurat@dr$tsne@cell.embeddings[selected,],pch=19,cex=0.35,col=cols,main=currentid)  \n",
    "    }\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### heatmap of DEG between healthy and diseas in each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DEG=readRDS(paste0(\"DE_healthyVSdisease_within_cluster_\",tag_celltype,\".Rds\"))\n",
    "#new naming\n",
    "names(DEG)=gp_name_replacing(names(DEG),c(paste0(\"CD8_TEM_effector_\",1:4),\"CD8_1\",\"CD8_2\",\"CD8_TCM_1\"),paste0(\"CD8_\",1:7)) \n",
    "#xlsx::write.xlsx(DEG[[\"CD4_1\"]], file=\"DE_healthyVSdisease_within_CD4_1.xlsx\", sheetName=\"CD4_1\", append=TRUE, row.names=T)\n",
    "ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(3,\"Set1\")\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "\n",
    "for(i in 1:length(DEG)){\n",
    "    if(sum(DEG[[i]]$DE)>1){\n",
    "        tagG=rownames(DEG[[i]])[DEG[[i]]$DE]\n",
    "        \n",
    "        healthy_cell_idx=which(Tcellsubcelltype_assignment==names(DEG)[i] & Exp_Seurat@meta.data$disease_assignment==1)\n",
    "        taghealthy=ExpressionMat_magic[healthy_cell_idx,tagG]\n",
    "        tmp1=fastcluster::hclust.vector(taghealthy,method='ward')\n",
    "        taghealthy=taghealthy[tmp1$order,]\n",
    "        \n",
    "        IBD_cell_idx=which(Tcellsubcelltype_assignment==names(DEG)[i] & Exp_Seurat@meta.data$disease_assignment==2)\n",
    "        tagIBD=ExpressionMat_magic[IBD_cell_idx,tagG]\n",
    "        tmp2=fastcluster::hclust.vector(tagIBD,method='ward')\n",
    "        tagIBD=tagIBD[tmp2$order,]\n",
    "        \n",
    "        \n",
    "        ColforPatient_ID=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(patient_type))\n",
    "\n",
    "        colcor=cbind(patient_ID=ColforPatient_ID[Exp_Seurat@meta.data$patient_assignment[c(healthy_cell_idx[tmp1$order],IBD_cell_idx[tmp2$order])]],\n",
    "                     IBD_status=c(rep(coltouse[2],nrow(taghealthy)),rep(coltouse[1],nrow(tagIBD))))\n",
    "        \n",
    "        \n",
    "        toplot=rbind(taghealthy,tagIBD)\n",
    "        toplot=t(scale(toplot))\n",
    "        toplot[toplot>1.5]=1.5\n",
    "        toplot[toplot<(-1.5)]=-1.5\n",
    "        pdf(paste0(\"heatmap_DEG_healthyVSdiseased_\",names(DEG)[i],\".pdf\"))\n",
    "        heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,ColSideColors=colcor,ColSideLabs=c(\"Patient ID\",\"IBD Status\"),\n",
    "                   labRow=rownames(toplot),labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none',main=names(DEG)[i])\n",
    "        dev.off()\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GWAS relationship"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "Group_type=unique(Tcellsubcelltype_assignment)\n",
    "Group_type=Group_type[gtools::mixedorder(Group_type)]\n",
    "\n",
    "get5k=Subsample_by_group(Tcellsubcelltype_assignment,5000)\n",
    "toplot=ExpressionMat_magic[get5k,colnames(ExpressionMat_magic) %in% GWAS_gene] #t(Exp_Seurat@data[unique(marker_genes$gene),get5k])\n",
    "\n",
    "toplot=t(scale(toplot))\n",
    "toplot[toplot>1.5]=1.5\n",
    "toplot[toplot<(-1.5)]=-1.5\n",
    "\n",
    "ColforDisease=RColorBrewer::brewer.pal(3,\"Set1\")[2:1]\n",
    "ColforSubCelltype=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(Group_type))\n",
    "ColforPatient_ID=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(patient_type))\n",
    "ColforTissue=RColorBrewer::brewer.pal(length(tissue_type),\"Set1\")\n",
    "ColSideColors=cbind(patient_ID=ColforPatient_ID[Exp_Seurat@meta.data$patient_assignment],\n",
    "                    tissue=ColforTissue[Exp_Seurat@meta.data$tissue_assignment],\n",
    "                    IBD_status=ColforDisease[Exp_Seurat@meta.data$disease_assignment],\n",
    "                    subcelltype=ColforSubCelltype[match(Tcellsubcelltype_assignment,Group_type)])\n",
    "ColSideColors=ColSideColors[get5k,]\n",
    "\n",
    "\n",
    "#ColSideColors can be a data.frame with each col indicate one color assignment\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "pdf(file=paste0(tag_celltype,\"_GWASbycluster_heatmap.pdf\"),width=15,height=10)\n",
    "heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,ColSideColors=ColSideColors,\n",
    "                   labRow=F,labCol=F,showColDendro=F,showRowDendro=F,scale='none')\n",
    "dev.off()\n",
    "\n",
    "png(file=paste0(tag_celltype,\"_GWASbycluster_heatmap.png\"),width=15,height=10,units = \"in\",res=150)\n",
    "heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,ColSideColors=ColSideColors,\n",
    "                   labRow=F,labCol=F,showColDendro=F,showRowDendro=F,scale='none')\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## B Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_Bcell_R2.Rds\")\n",
    "load(\"Ig_type.Rdata\")\n",
    "tag_celltype=\"Bcell\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#cell type assignment\n",
    "celltype_assignment_df=data.frame(cluster=c(\"17,16\",\"8,2\",\"13,5\",\"6,0\",\"7,10\",\"11,1\",\"3\",\"18,12,9,19\",\"21,20\",\"4,14,15\"),\n",
    "                                  celltype=c(\"Naive_B\",\"Memory_B\",\"Plasma\",\"NKT\",\"CD8\",\"CD4\",\"gdT\",\"NK\",\"unknown\",\"M_phi\"),\n",
    "                                  stringsAsFactors=F)\n",
    "\n",
    "celltype_assignment_expanded=cellTypeDF_processing(celltype_assignment_df)\n",
    "\n",
    "celltype_assignment=gp_name_replacing(as.character(Exp_Seurat@meta.data$global_gps),celltype_assignment_expanded$cluster,celltype_assignment_expanded$celltype)\n",
    "subcelltype_assignment=gp_name_replacing(as.character(Exp_Seurat@meta.data$global_gps),celltype_assignment_expanded$cluster,celltype_assignment_expanded$subcelltype)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#cell type assignment\n",
    "Bcell_assignment_df=data.frame(cluster=c(\"6,14,5,2,13,10,12\",\"3,0,1,8\",\"11\",\"4,7,9\"),\n",
    "                                  celltype=c(\"Plasma\",\"Memory_B\",\"Breg\",\"Naive_B\"),\n",
    "                                  stringsAsFactors=F)\n",
    "Bcell_assignment_expanded=cellTypeDF_processing(Bcell_assignment_df)\n",
    "Bcell_assignment=gp_name_replacing(as.character(Exp_Seurat@ident),Bcell_assignment_expanded$cluster,Bcell_assignment_expanded$celltype)\n",
    "Bcellsubcelltype_assignment=gp_name_replacing(as.character(Exp_Seurat@ident),Bcell_assignment_expanded$cluster,Bcell_assignment_expanded$subcelltype)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### cell percentage of each cell type in tissue/patient/disease_status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample_cell_num=readRDS(file=\"sample_cell_num.rds\")\n",
    "\n",
    "tmp=table(Bcellsubcelltype_assignment,Exp_Seurat@meta.data$sample_gp)\n",
    "tmp_rownames=rownames(tmp)\n",
    "colnames(tmp)=sample_list[as.numeric(colnames(tmp))]\n",
    "tmp_percent_by_celltype=tmp\n",
    "tmp_percent_by_all=tmp\n",
    "csum=colSums(tmp)\n",
    "for(i in 1:nrow(tmp)){\n",
    "    tmp_percent_by_celltype[i,]=tmp[i,]/csum\n",
    "    tmp_percent_by_all[i,]=tmp[i,]/sample_cell_num[match(colnames(tmp),names(sample_cell_num))]\n",
    "}\n",
    "tmp=data.table(as.data.frame.matrix(tmp))\n",
    "tmp_percent_by_celltype=data.table(as.data.frame.matrix(tmp_percent_by_celltype))\n",
    "tmp_percent_by_all=data.table(as.data.frame.matrix(tmp_percent_by_all))\n",
    "tmp$majorcelltype=tmp_rownames\n",
    "tmp_percent_by_celltype$majorcelltype=tmp_rownames\n",
    "tmp_percent_by_all$majorcelltype=tmp_rownames\n",
    "\n",
    "tmp=melt(tmp, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"count\")\n",
    "tmp_percent_by_celltype=melt(tmp_percent_by_celltype, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp_percent_by_celltype),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"percent_by_celltype\")\n",
    "tmp_percent_by_all=melt(tmp_percent_by_all, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp_percent_by_all),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"percent_by_all\")\n",
    "tmp$percent_by_celltype=tmp_percent_by_celltype$percent_by_celltype\n",
    "tmp$percent_by_all=tmp_percent_by_all$percent_by_all\n",
    "tmp$patient=gsub(\"_.*\",\"\",tmp$batch)\n",
    "tmp$tissue=gsub(\"^.*_\",\"\",tmp$batch)\n",
    "tmp$disease_status=gp_name_replacing(substr(tmp$batch, 1, 1),c(\"C\",\"U\"),c(\"Healthy\",\"IBD\"))\n",
    "tmp=data.frame(tmp)\n",
    "\n",
    "Group_type=unique(Bcellsubcelltype_assignment)\n",
    "Group_type=Group_type[gtools::mixedorder(Group_type)]\n",
    "\n",
    "pdf(paste0(tag_celltype,\" subtype distribution.pdf\"),width=12,height=5)\n",
    "\n",
    "coltouse=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(Group_type))\n",
    "#options(repr.plot.width=12, repr.plot.height=5)\n",
    "ggplot() +\n",
    "geom_bar(aes(y = count, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"count\") + labs(y = \"count of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = count, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"stack\") +\n",
    "scale_y_continuous() +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"count\") + labs(y = \"count of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_celltype, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by cell type\") + labs(y = paste0(\"pecentage of cells in given batch's \",tag_celltype),fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_all, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by all cells\") + labs(y = \"pecentage of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_all, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"stack\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by all cells\") + labs(y = \"pecentage of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "dev.off()\n",
    "\n",
    "# define the summary function\n",
    "f <- function(x) {\n",
    "  r <- quantile(x, probs = c(0.1, 0.25, 0.5, 0.75, 0.9))\n",
    "  names(r) <- c(\"ymin\", \"lower\", \"middle\", \"upper\", \"ymax\")\n",
    "  r\n",
    "}\n",
    "\n",
    "o <- function(x) {\n",
    "  subset(x, x < quantile(x)[2] | quantile(x)[4] < x)\n",
    "}\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(3,\"Set1\")\n",
    "\n",
    "options(warn=-1) #turn off warning\n",
    "pdf(paste0(\"healthy vs IBD in \",tag_celltype,\" subtype.pdf\"),width=25,height=12)\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=count, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"count of cells in given batch\") + ggtitle(\"count\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=percent_by_celltype, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"pecentage of cells in given batch\") + ggtitle(\"percent by cell type\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=percent_by_all, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"pecentage of cells in given batch\") + ggtitle(\"percent by all cells\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "dev.off()\n",
    "options(warn=0) #turn on warning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### gene-group heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#contrust matrix, each row is a state\n",
    "ExpressionMat_by_gp=cbind(Exp_Seurat@ident,2^t(Exp_Seurat@data))\n",
    "colnames(ExpressionMat_by_gp)[1]=\"states\"\n",
    "tagV=c(\"TRAV\",\"TRAC\",\"TRBV\",\"TRBC\",\"TRGV\",\"TRGC\",\"TRDV\",\"TRDC\",\n",
    "       \"IGHV\",\"IGHG\",\"IGHA\",\"IGHD\",\"IGHM\",\"IGHE\",\"IGLV\",\"IGLL\",\"IGKV\",\"IGLC\",\"IGKC\")\n",
    "togetrid=c(\"IGHMBP2\")\n",
    "for(i in 1:length(tagV)){\n",
    "    tagG=grep(paste0(\"^\",tagV[i]),colnames(ExpressionMat_by_gp))\n",
    "    tagG=setdiff(tagG,togetrid)\n",
    "    if(length(tagG)==0){\n",
    "        next\n",
    "    }\n",
    "    tmp=ExpressionMat_by_gp[,tagG]\n",
    "    if(!is.null(ncol(tmp))){\n",
    "        tmp=rowSums(tmp)\n",
    "    }\n",
    "    ExpressionMat_by_gp=cbind(ExpressionMat_by_gp,tmp)\n",
    "    colnames(ExpressionMat_by_gp)[ncol(ExpressionMat_by_gp)]=tagV[i]\n",
    "}\n",
    "ExpressionMat_by_gp=data.table(ExpressionMat_by_gp)\n",
    "ExpressionMat_by_gp=ExpressionMat_by_gp[,lapply(.SD, mean), by=states]\n",
    "ExpressionMat_by_gp=ExpressionMat_by_gp[order(ExpressionMat_by_gp$states)]\n",
    "ExpressionMat_by_gp[,states:=NULL]\n",
    "ExpressionMat_by_gp=data.frame(ExpressionMat_by_gp)\n",
    "colnames(ExpressionMat_by_gp)=gsub(\"\\\\.\",\"-\",colnames(ExpressionMat_by_gp))\n",
    "ExpressionMat_by_gp=log2(ExpressionMat_by_gp)\n",
    "rownames(ExpressionMat_by_gp)=paste0(\"gp_\",levels(Exp_Seurat@ident))\n",
    "write.table(t(ExpressionMat_by_gp),paste0(\"mediam_gene_expression_\",tag_celltype,\".tsv\"),sep=\"\\t\",col.names=T,row.names=T,quote=F)\n",
    "saveRDS(ExpressionMat_by_gp,file=paste0(\"ExpressionMat_by_gp_\",tag_celltype,\".Rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SLC22A2, CCNE1, MME, FLT3, TNFRSF8, CD80, ACKR3, IL10, EBI3\n"
     ]
    }
   ],
   "source": [
    "ExpressionMat_by_gp=readRDS(paste0(\"ExpressionMat_by_gp_\",tag_celltype,\".Rds\"))\n",
    "#B cell\n",
    "taggene=list()\n",
    "taggene=c(\"SLC22A2\",\"CD93\",\"BACH2\",\"BCL6\",\"BHLHE41\",\"CCNE1\",\"CXCR4\",\"MME\",\"FLT3\",\"SDC1\",\"CD19\",\"CD1d\",\"MS4A1\",\"CR2\",\"CD22\",\n",
    "          \"FCER2\",\"CD24\",\"TNFRSF17\",\"CD27\",\"CD3G\",\"TNFRSF8\",\"PECAM1\",\"CR1\",\"CD38\",\"CD40\",\"SPN\",\"CD5\",\"CD69\",\"CD80\",\"CD86\",\"CD9\"\n",
    "          ,\"CXCR5\",\"ACKR3\",\"TCF3\",\"EBF1\",\"FAS\",\"HLA-DMA\",\"HLA-DOA\",\"HLA-DOB\",\"HLA-DPA1\",\"HLA-DQA1\",\"HLA-DRA\",\"IgHA\",\"IGHD\",\"IgHG\",\n",
    "          \"IGHM\",\"IL10\",\"IL2RA\",\"EBI3\",\"IRF4\",\"IRF8\",\"Notch2\",\"Pax5\",\"POU2AF1\",\"Prdm1\",\"PTPRC\",\"SPI1\",\"SDC1\",\"SLAMF1\",\n",
    "          \"TGFB1\",\"XBP1\")\n",
    "\n",
    "taggene=toupper(taggene)\n",
    "tagExpressionNormed=ExpressionMat_by_gp[,which(colnames(ExpressionMat_by_gp) %in% taggene)]\n",
    "#tagExpressionNormed=tagExpressionNormed[,order(colnames(tagExpressionNormed))]\n",
    "missing_gene=paste0(taggene[!taggene %in% colnames(ExpressionMat_by_gp)],collapse = \", \")\n",
    "message(missing_gene)\n",
    "\n",
    "rownames(tagExpressionNormed)=Bcell_assignment_expanded$subcelltype[order(as.numeric(Bcell_assignment_expanded$cluster))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=8)\n",
    "\n",
    "pdf(paste0(\"clusterGeneExp_\",tag_celltype,\"_heatmap.pdf\"),width=15, height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "heatmap.2(scale(tagExpressionNormed),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"cell assignment vs. gene\\nmissing:\",missing_gene),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "\n",
    "heatmap.2(as.matrix(tagExpressionNormed),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"cell assignment vs. gene\\nmissing:\",missing_gene),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### VDJ vs RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0(\"VDJassignment_vs_RNAassignment_\",tag_celltype,\"_heatmap.pdf\"),width=6,height=9)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "tmp=table(Bcellsubcelltype_assignment,Exp_Seurat@meta.data$cell_assignment)\n",
    "\n",
    "heatmap.2(t(apply(tmp,1,function(x){x/sum(x)})),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=F,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"RNA-based cell type assignment\\n vs. VDJ based\"),key.title=\"\",key.xlab=\"\",margins=c(15,10),cellnote=tmp)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### DEG within groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tag_cluster=Bcell_assignment_expanded$cluster[Bcell_assignment_expanded$celltype %in% c(\"Plasma\")]\n",
    "cluster2rm=setdiff(celltype_assignment_expanded$cluster,tag_cluster)\n",
    "Exp_Seurat_B = filter_ident(Exp_Seurat,cluster2rm)\n",
    "marker_genes=FindAllMarkers(object = Exp_Seurat_B, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)\n",
    "marker_genes$cluster=gp_name_replacing(as.character(marker_genes$cluster),Bcell_assignment_expanded$cluster,Bcell_assignment_expanded$subcelltype)\n",
    "write_DEG(marker_genes,\"Plasma\")\n",
    "\n",
    "tag_cluster=Bcell_assignment_expanded$cluster[Bcell_assignment_expanded$celltype %in% c(\"Memory_B\")]\n",
    "cluster2rm=setdiff(celltype_assignment_expanded$cluster,tag_cluster)\n",
    "Exp_Seurat_B = filter_ident(Exp_Seurat,cluster2rm)\n",
    "marker_genes=FindAllMarkers(object = Exp_Seurat_B, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)\n",
    "marker_genes$cluster=gp_name_replacing(as.character(marker_genes$cluster),Bcell_assignment_expanded$cluster,Bcell_assignment_expanded$subcelltype)\n",
    "write_DEG(marker_genes,\"Memory_B\")\n",
    "\n",
    "tag_cluster=Bcell_assignment_expanded$cluster[Bcell_assignment_expanded$celltype %in% c(\"Naive_B\")]\n",
    "cluster2rm=setdiff(celltype_assignment_expanded$cluster,tag_cluster)\n",
    "Exp_Seurat_B = filter_ident(Exp_Seurat,cluster2rm)\n",
    "marker_genes=FindAllMarkers(object = Exp_Seurat_B, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)\n",
    "marker_genes$cluster=gp_name_replacing(as.character(marker_genes$cluster),Bcell_assignment_expanded$cluster,Bcell_assignment_expanded$subcelltype)\n",
    "write_DEG(marker_genes,\"Naive_B\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Ig infomation mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "poslist=list(IGApos,IGDpos,IGEpos,IGGpos,IGMpos)\n",
    "names(poslist)=c(\"IGA\",\"IGD\",\"IGE\",\"IGG\",\"IGM\")\n",
    "Exp_Seurat@meta.data$IgH_label=\"unknown\"\n",
    "existing_tag=c()\n",
    "for(i in 1:length(poslist)){\n",
    "    current_tag=which(rownames(Exp_Seurat@meta.data) %in% poslist[[i]])\n",
    "    overlap_tag=intersect(current_tag,existing_tag)\n",
    "    Exp_Seurat@meta.data$IgH_label[current_tag]=names(poslist)[i]\n",
    "    Exp_Seurat@meta.data$IgH_label[overlap_tag]=\"multiple_C_genes\"\n",
    "    existing_tag=union(existing_tag,current_tag) \n",
    "}\n",
    "\n",
    "poslist=list(IGLpos,IGKpos)\n",
    "names(poslist)=c(\"IGL\",\"IGK\")\n",
    "Exp_Seurat@meta.data$IgL_label=\"unknown\"\n",
    "existing_tag=c()\n",
    "for(i in 1:length(poslist)){\n",
    "    current_tag=which(rownames(Exp_Seurat@meta.data) %in% poslist[[i]])\n",
    "    overlap_tag=intersect(current_tag,existing_tag)\n",
    "    Exp_Seurat@meta.data$IgL_label[current_tag]=names(poslist)[i]\n",
    "    Exp_Seurat@meta.data$IgL_label[overlap_tag]=\"multiple_C_genes\"\n",
    "    existing_tag=union(existing_tag,current_tag) \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in plotByGroup(Exp_Seurat, Group_assignment = expansion_info_summ$expand_lvl[match(colnames(Exp_Seurat@data), :\n",
      "“NAs in Group_assignment”"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "position_index_matrix=readRDS(\"B_cell_position_index_matrix.rds\")\n",
    "expansion_info_summ=data.table(position_index_matrix[!is.na(position_index_matrix$celltype),c(\"barcode\",\"expand_lvl\")])\n",
    "expansion_info_summ=expansion_info_summ[,.(expand_lvl=max(expand_lvl)),by=barcode]\n",
    "expansion_info_summ$expand_lvl[expansion_info_summ$expand_lvl>10]=\"10+\"\n",
    "\n",
    "pdf(paste0(\"TSNE_mito_\",tag_celltype,\".pdf\"),width=18,height=15)\n",
    "par(mfrow=c(3,3))\n",
    "plotByGroup(Exp_Seurat,Group_assignment=sample_list[Exp_Seurat@meta.data$sample_gp],\n",
    "            Group_type=sample_list,main=\"Batchs\") #batchs\n",
    "plotByGroup(Exp_Seurat,Group_assignment=tissue_type[Exp_Seurat@meta.data$tissue_assignment],\n",
    "            Group_type=tissue_type,main=\"tissue type\") #tissue\n",
    "plotByGroup(Exp_Seurat,Group_assignment=patient_type[Exp_Seurat@meta.data$patient_assignment],\n",
    "            Group_type=patient_type,main=\"patient id\") #patient id\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@meta.data$cell_assignment,\n",
    "            Group_type=c('failed BCR/TCR','B cell','T cell','dual-label','unknown'),main=\"B/T classification based on VDJ\",backGroundGroup=\"unknown\") #BT\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Bcellsubcelltype_assignment,main=\"B cell substype\",Addtext=T) \n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@meta.data$IgH_label,main=\"IgH type\",Addtext=T,\n",
    "            Group_type = c(\"IGA\",\"IGD\",\"IGE\",\"IGG\",\"IGM\",\"multiple_C_genes\",\"unknown\"),backGroundGroup=\"unknown\") \n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@meta.data$IgL_label,main=\"IgL type\",Addtext=T,\n",
    "            Group_type = c(\"IGL\",\"IGK\",\"multiple_C_genes\",\"unknown\"),backGroundGroup=\"unknown\") \n",
    "plotByGroup(Exp_Seurat,Group_assignment=c(\"healthy\",\"diseased\")[Exp_Seurat@meta.data$disease_assignment],\n",
    "            Group_type=c(\"diseased\",\"healthy\"),main=\"disease condition\") #patient id\n",
    "plotByGroup(Exp_Seurat,Group_assignment=expansion_info_summ$expand_lvl[match(colnames(Exp_Seurat@data),expansion_info_summ$barcode)],\n",
    "            Group_type=c(1:9,\"10+\"),main=\"expansion level\",tagcol=colorRampPalette(rev(RColorBrewer::brewer.pal(7,\"RdBu\")))(10)) \n",
    "#plotByScore(Exp_Seurat,Score_assignment=Exp_Seurat@meta.data$S.Score,main=\"\") \n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "clusterVsPatient=table(Bcellsubcelltype_assignment,Exp_Seurat@meta.data$sample_gp)\n",
    "colnames(clusterVsPatient)=sample_list[as.numeric(colnames(clusterVsPatient))]\n",
    "clusterVsPatient_colPercent=round(apply(clusterVsPatient,2,function(x){x/sum(x)}),2)\n",
    "clusterVsPatient_rowPercent=round(t(apply(clusterVsPatient,1,function(x){x/sum(x)})),2)\n",
    "write.table(clusterVsPatient,file=paste0(\"clusterVsPatient_table_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(clusterVsPatient_rowPercent,file=paste0(\"clusterVsPatient_ClusterPercentage_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(clusterVsPatient_colPercent,file=paste0(\"clusterVsPatient_PatientPercentage_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "clusterVsPatient_rowPercent[clusterVsPatient_rowPercent>0.2]=0.2\n",
    "clusterVsPatient_colPercent[clusterVsPatient_colPercent>0.5]=0.5\n",
    "\n",
    "pdf(paste0(\"clusterSample_\",tag_celltype,\"_heatmap.pdf\"),width=15, height=8)\n",
    "heatmap.2(clusterVsPatient_colPercent,col=colorRampPalette(col2),\n",
    "          lhei=c(1,5),lwid=c(0.1,5),margins = c(10,10),trace=\"none\",density.info=\"none\",dendrogram='col',Colv=T,Rowv=T,\n",
    "          notecol=\"black\",key=F,cellnote=clusterVsPatient)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### heatmap cell-DEG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "Group_type=unique(Bcellsubcelltype_assignment)\n",
    "Group_type=Group_type[gtools::mixedorder(Group_type)]\n",
    "\n",
    "marker_genes=readRDS(paste0(\"markerGenes_\",base_Rdata_name,\"_\",tag_celltype,\".rds\"))\n",
    "get5k=Subsample_by_group(Bcellsubcelltype_assignment,5000)\n",
    "toplot=ExpressionMat_magic[get5k,unique(marker_genes$gene)] #t(Exp_Seurat@data[unique(marker_genes$gene),get5k])\n",
    "\n",
    "tmp=as.numeric(gp_name_replacing(Bcellsubcelltype_assignment[get5k],Group_type,1:length(Group_type)))\n",
    "globall_ordering=fastcluster::hclust.vector(toplot,method='ward')\n",
    "globall_sorting_idx=match(1:length(get5k),globall_ordering$order)\n",
    "colOrder=order(tmp,Exp_Seurat@meta.data$disease_assignment[get5k],globall_sorting_idx)\n",
    "\n",
    "toplot=toplot[colOrder,]\n",
    "toplot=t(scale(toplot))\n",
    "toplot[toplot>1.5]=1.5\n",
    "toplot[toplot<(-1.5)]=-1.5\n",
    "\n",
    "ColforDisease=RColorBrewer::brewer.pal(3,\"Set1\")[2:1]\n",
    "ColforSubCelltype=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(Group_type))\n",
    "ColforPatient_ID=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(patient_type))\n",
    "ColSideColors=cbind(patient_ID=ColforPatient_ID[Exp_Seurat@meta.data$patient_assignment],\n",
    "                    disease=ColforDisease[Exp_Seurat@meta.data$disease_assignment],\n",
    "                    subcelltype=ColforSubCelltype[match(Bcellsubcelltype_assignment,Group_type)])\n",
    "ColSideColors=ColSideColors[get5k[colOrder],]\n",
    "\n",
    "RowSideColors=matrix(\"white\",nrow(toplot),max(as.numeric(marker_genes$cluster)))\n",
    "colnames(RowSideColors)=Group_type\n",
    "for(i in 1:max(as.numeric(marker_genes$cluster))){\n",
    "    tagG=marker_genes$gene[marker_genes$cluster==i-1]\n",
    "    tagsubtype=Bcell_assignment_expanded$subcelltype[Bcell_assignment_expanded$cluster==i-1]\n",
    "    RowSideColors[rownames(toplot) %in% tagG,which(Group_type==tagsubtype)]=ColforSubCelltype[which(Group_type==tagsubtype)]\n",
    "}\n",
    "\n",
    "#ColSideColors can be a data.frame with each col indicate one color assignment\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "pdf(file=paste0(tag_celltype,\"_DEGbycluster_heatmap.pdf\"),width=15,height=10)\n",
    "heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,RowSideColors=RowSideColors,ColSideColors=ColSideColors,\n",
    "                   labRow=F,labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none')\n",
    "dev.off()\n",
    "\n",
    "png(file=paste0(tag_celltype,\"_DEGbycluster_heatmap.png\"),width=15,height=10,units = \"in\",res=150)\n",
    "heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,RowSideColors=RowSideColors,ColSideColors=ColSideColors,\n",
    "                   labRow=F,labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none')\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### per gene plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "require(feather)\n",
    "#MAGIC\n",
    "forMagic=\"For_magic.feather\"\n",
    "afterMagic=\"Magic_out.feather\"\n",
    "\n",
    "temp_Exp=as.data.frame(t(Exp_Seurat@data))\n",
    "#magic\n",
    "write_feather(temp_Exp, forMagic)\n",
    "system(paste0(\"python /home/ahe/tools/Lightbulb/src/Lightbulb_MagicWrapper.py --matx \",forMagic,\" --out \",afterMagic))\n",
    "ExpressionMat_magic = data.matrix(read_feather(afterMagic))\n",
    "colnames(ExpressionMat_magic)=gsub(\"MAGIC \", \"\", colnames(ExpressionMat_magic))\n",
    "\n",
    "saveRDS(ExpressionMat_magic,file=paste0(tag_celltype,\"_magic.rds\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SLC22A2, CCNE1, MME, TNFRSF8, CD80, ACKR3, FLT3, IGHA, IGHG, IL10, EBI3\n",
      "\n",
      "IGHA, IGHG, IGHD , CCNE1, ACKR3, CD80\n",
      "FLT3\n",
      "\n",
      "\n",
      "MME, SLC22A2\n",
      "TNFRSF8, CD80, FLT3\n",
      "MME\n",
      "IL10, EBI3\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ExpressionMat_magic=readRDS(paste0(tag_celltype,\"_magic.rds\"))\n",
    "taggene=list()\n",
    "taggene[[1]]=c(\"SLC22A2\",\"CD93\",\"BACH2\",\"BCL6\",\"BHLHE41\",\"CCNE1\",\"CXCR4\",\"MME\",\"SDC1\",\"CD19\",\"CD1d\",\"MS4A1\",\"CR2\",\"CD22\",\"FCER2\",\n",
    "               \"CD24\",\"TNFRSF17\",\"CD27\",\"CD3G\",\"TNFRSF8\",\"PECAM1\",\"CR1\",\"CD38\",\"CD40\",\"SPN\",\"CD5\",\"CD69\",\"CD80\",\"CD86\",\"CD9\",\"CXCR4\",\n",
    "               \"CXCR5\",\"ACKR3\",\"TCF3\",\"EBF1\",\"FAS\",\"FLT3\",\"HLA-DMA\",\"HLA-DOA\",\"HLA-DOB\",\"HLA-DPA1\",\"HLA-DQA1\",\"HLA-DRA\",\"IgHA\",\n",
    "               \"IGHD\",\"IgHG\",\"IGHM\",\"IL10\",\"IL2RA\",\"EBI3\",\"IRF4\",\"IRF8\",\"Notch2\",\"Pax5\",\"POU2AF1\",\"Prdm1\",\"PTPRC\",\"SPI1\",\"SDC1\"\n",
    "               ,\"SLAMF1\",\"TGFB1\",\"TNFRSF17\",\"XBP1\")\n",
    "taggene[[2]]=c(\"CD19\",\"MS4A1\",\"FCER2\",\"CD38\",\"CD40\",\"SLAMF1\",\"IGHD\",\"CD27\")\n",
    "taggene[[3]]=c(\"BACH2\",\"IRF8\",\"Pax5\",\"SPI1\",\"CD19\",\"MS4A1\",\"FCER2\",\"CD27\",\"CD38\",\"CD40\",\"SLAMF1\",\"IgHA\",\"IgHG\",\"CD27\",\"IGHD \",\"POU2AF1\",\"CCNE1\",\"CXCR4\",\"CXCR5\",\"ACKR3\",\"FAS\",\"CD80\")\n",
    "taggene[[4]]=c(\"Prdm1\",\"IRF4\",\"XBP1\",\"CD9\",\"CD19\",\"MS4A1\",\"CD27\",\"PECAM1\",\"CD38\",\"CD40\",\"FAS\",\"CXCR4\",\"MS4A1\",\"CD38\",\"SDC1\",\"IGHM\",\"IGHD\",\"CXCR4\",\"TNFRSF17\",\"HLA-DPA1\",\"FLT3\",\"HLA-DMA\",\"HLA-DOA\",\"HLA-DOB\",\"HLA-DQA1\",\"HLA-DRA\",\"BCL6\")\n",
    "taggene[[5]]=c(\"Prdm1\",\"IRF4\",\"XBP1\",\"CXCR4\",\"CD27\",\"CD38\",\"SDC1\",\"TNFRSF17\",\"CD19\",\"MS4A1\",\"HLA-DRB1\",\"IGHD\")\n",
    "taggene[[6]]=c(\"CD24\",\"CD38\",\"CD19\")\n",
    "taggene[[7]]=c(\"CD19\",\"MS4A1\",\"CR2\",\"CD22\",\"FCER2\",\"CD24\",\"MME\",\"CD27\",\"CD38\",\"CD24\",\"CR1\",\"Pax5\",\"EBF1\",\"TCF3\",\"SPI1\",\"SLC22A2\")\n",
    "taggene[[8]]=c(\"CD27\",\"CD19\",\"MS4A1\",\"IL2RA\",\"TNFRSF8\",\"CD69\",\"CD80\",\"CD86\",\"FLT3\")\n",
    "taggene[[9]]=c(\"MME\",\"CD19\",\"MS4A1\",\"FCER2\",\"CD27\",\"CD38\",\"TNFRSF17\",\"CD24\",\"Bcl6\",\"FAS\")\n",
    "taggene[[10]]=c(\"CD1d\",\"CD4\",\"CD19\",\"CR2\",\"CD24\",\"CD27\",\"IGHM\",\"IL10\",\"EBI3\",\"TGFB1\",\"CD38\")\n",
    "taggene[[11]]=c(\"IGHM\",\"CR2\",\"CR1\",\"CD93\")\n",
    "taggene[[12]]=c(\"IGHM\",\"CR2\",\"CR1\",\"FCER2\",\"CD93\",\"Notch2\")\n",
    "taggene[[13]]=c(\"FCER2\",\"CD19\",\"SPN\",\"PTPRC\",\"CD5\",\"BHLHE41\")\n",
    "taggene[[14]]=c(\"FCER2\",\"CD19\",\"SPN\",\"PTPRC\",\"CD5\")\n",
    "names(taggene)=c(\"all\",\"Naive\",\"Memory\",\"LongLivePlasma\",\"ShortLivePlasma\",\n",
    "                 \"Transitional\",\"Follicular\",\"Activated\",\"GerminalCenter\",\"Breg\",\n",
    "                 \"B1B\",\"MariginalZone\",\"B1aB\",\"B1bB\")\n",
    "missing_gene=list()\n",
    "tagExpressionNormed=list()\n",
    "for(i in 1:length(taggene)){\n",
    "    taggene[[i]]=toupper(taggene[[i]])\n",
    "    tagExpressionNormed[[i]]=ExpressionMat_magic[,which(colnames(ExpressionMat_magic) %in% taggene[[i]])]\n",
    "    tagExpressionNormed[[i]]=tagExpressionNormed[[i]][,order(colnames(tagExpressionNormed[[i]]))]\n",
    "    missing_gene[[i]]=paste0(taggene[[i]][!taggene[[i]] %in% colnames(tagExpressionNormed[[i]])],collapse = \", \")\n",
    "    message(missing_gene[[i]])\n",
    "}\n",
    "\n",
    "#SLC22A2, CCNE1, MME, FLT3, TNFRSF8, CD80, ACKR3, IL10, EBI3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "rbPal <- colorRampPalette(c('gray','#2b83ba','#abdda4','#ffffbf','#fdae61','#d7191c'))\n",
    "selected=sample(1:nrow(tagExpressionNormed[[1]]),2000)\n",
    "for(x in 1:length(taggene)){\n",
    "    pdf(paste0(tag_celltype,\"_\",names(taggene)[x],\"_per gene mito TSNE_magic.pdf\"),width=22,height=15)\n",
    "    par(mfrow=c(4,6))\n",
    "    for(i in 1:ncol(tagExpressionNormed[[x]])){\n",
    "        currentid=colnames(tagExpressionNormed[[x]])[i]\n",
    "        #calculate breaks\n",
    "        lowerbound=sort(tagExpressionNormed[[x]][selected,i])[round(length(selected)/33)]\n",
    "        upperbound=sort(tagExpressionNormed[[x]][selected,i])[round(length(selected)/33*32)]\n",
    "        if(upperbound!=lowerbound){\n",
    "            breaks=c(min(tagExpressionNormed[[x]][selected,i])-1,seq(lowerbound, upperbound, length.out=25),max(tagExpressionNormed[[x]][selected,i])+1)\n",
    "            cols=rbPal(25)[as.numeric(cut(tagExpressionNormed[[x]][selected,i],breaks = breaks, include.lowest=TRUE))]\n",
    "        }else{\n",
    "            breaks=c(min(tagExpressionNormed[[x]][selected,i])-1,max(tagExpressionNormed[[x]][selected,i])/2,max(tagExpressionNormed[[x]][selected,i]))\n",
    "            cols=rbPal(2)[as.numeric(cut(tagExpressionNormed[[x]][selected,i],breaks = breaks, include.lowest=TRUE))]\n",
    "        }\n",
    "        plot(Exp_Seurat@dr$tsne@cell.embeddings[selected,],pch=19,cex=0.35,col=cols,main=currentid)  \n",
    "    }\n",
    "    dev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### heatmap of DEG between healthy and diseas in each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DEG=readRDS(paste0(\"DE_healthyVSdisease_within_cluster_\",tag_celltype,\".Rds\"))\n",
    "ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(3,\"Set1\")\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "\n",
    "for(i in 1:length(DEG)){\n",
    "    if(sum(DEG[[i]]$DE)>1){\n",
    "        tagG=rownames(DEG[[i]])[DEG[[i]]$DE]\n",
    "        taghealthy=ExpressionMat_magic[Bcellsubcelltype_assignment==names(DEG)[i] & Exp_Seurat@meta.data$disease_assignment==1,tagG]\n",
    "        tmp=fastcluster::hclust.vector(taghealthy,method='single')\n",
    "        taghealthy=taghealthy[tmp$order,]\n",
    "        tagIBD=ExpressionMat_magic[Bcellsubcelltype_assignment==names(DEG)[i] & Exp_Seurat@meta.data$disease_assignment==2,tagG]\n",
    "        tmp=fastcluster::hclust.vector(tagIBD,method='single')\n",
    "        tagIBD=tagIBD[tmp$order,]\n",
    "        colcor=c(rep(coltouse[2],nrow(taghealthy)),rep(coltouse[1],nrow(tagIBD)))\n",
    "        toplot=rbind(taghealthy,tagIBD)\n",
    "        toplot=t(scale(toplot))\n",
    "        toplot[toplot>1.5]=1.5\n",
    "        toplot[toplot<(-1.5)]=-1.5\n",
    "        pdf(paste0(\"heatmap_DEG_healthyVSdiseased_\",names(DEG)[i],\".pdf\"))\n",
    "        heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,ColSideColors=colcor,ColSideLabs=\"IBD Status\",\n",
    "                   labRow=rownames(toplot),labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none',main=names(DEG)[i])\n",
    "        dev.off()\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NK Cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_NKcell_R2.Rds\")\n",
    "tag_celltype=\"NKcell\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#cell type assignment\n",
    "celltype_assignment_df=data.frame(cluster=c(\"17,16\",\"8,2\",\"13,5\",\"6,0\",\"7,10\",\"11,1\",\"3\",\"18,12,9,19\",\"21,20\",\"4,14,15\"),\n",
    "                                  celltype=c(\"Naive_B\",\"Memory_B\",\"Plasma\",\"NKT\",\"CD8\",\"CD4\",\"gdT\",\"NK\",\"unknown\",\"M_phi\"),\n",
    "                                  stringsAsFactors=F)\n",
    "\n",
    "celltype_assignment_expanded=cellTypeDF_processing(celltype_assignment_df)\n",
    "\n",
    "celltype_assignment=gp_name_replacing(as.character(Exp_Seurat@meta.data$global_gps),celltype_assignment_expanded$cluster,celltype_assignment_expanded$celltype)\n",
    "subcelltype_assignment=gp_name_replacing(as.character(Exp_Seurat@meta.data$global_gps),celltype_assignment_expanded$cluster,celltype_assignment_expanded$subcelltype)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### cell percentage of each cell type in tissue/patient/disease_status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample_cell_num=readRDS(file=\"sample_cell_num.rds\")\n",
    "\n",
    "tmp=table(paste0(\"NK_\",Exp_Seurat@ident),Exp_Seurat@meta.data$sample_gp)\n",
    "tmp_rownames=rownames(tmp)\n",
    "colnames(tmp)=sample_list[as.numeric(colnames(tmp))]\n",
    "tmp_percent_by_celltype=tmp\n",
    "tmp_percent_by_all=tmp\n",
    "csum=colSums(tmp)\n",
    "for(i in 1:nrow(tmp)){\n",
    "    tmp_percent_by_celltype[i,]=tmp[i,]/csum\n",
    "    tmp_percent_by_all[i,]=tmp[i,]/sample_cell_num[match(colnames(tmp),names(sample_cell_num))]\n",
    "}\n",
    "tmp=data.table(as.data.frame.matrix(tmp))\n",
    "tmp_percent_by_celltype=data.table(as.data.frame.matrix(tmp_percent_by_celltype))\n",
    "tmp_percent_by_all=data.table(as.data.frame.matrix(tmp_percent_by_all))\n",
    "tmp$majorcelltype=tmp_rownames\n",
    "tmp_percent_by_celltype$majorcelltype=tmp_rownames\n",
    "tmp_percent_by_all$majorcelltype=tmp_rownames\n",
    "\n",
    "tmp=melt(tmp, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"count\")\n",
    "tmp_percent_by_celltype=melt(tmp_percent_by_celltype, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp_percent_by_celltype),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"percent_by_celltype\")\n",
    "tmp_percent_by_all=melt(tmp_percent_by_all, id.vars = \"majorcelltype\",measure.vars = setdiff(colnames(tmp_percent_by_all),\"majorcelltype\"),\n",
    "         variable.name=\"batch\",value.name=\"percent_by_all\")\n",
    "tmp$percent_by_celltype=tmp_percent_by_celltype$percent_by_celltype\n",
    "tmp$percent_by_all=tmp_percent_by_all$percent_by_all\n",
    "tmp$patient=gsub(\"_.*\",\"\",tmp$batch)\n",
    "tmp$tissue=gsub(\"^.*_\",\"\",tmp$batch)\n",
    "tmp$disease_status=gp_name_replacing(substr(tmp$batch, 1, 1),c(\"C\",\"U\"),c(\"Healthy\",\"IBD\"))\n",
    "tmp=data.frame(tmp)\n",
    "\n",
    "Group_type=unique(paste0(\"NK_\",Exp_Seurat@ident))\n",
    "Group_type=Group_type[gtools::mixedorder(Group_type)]\n",
    "\n",
    "pdf(paste0(tag_celltype,\" subtype distribution.pdf\"),width=12,height=5)\n",
    "\n",
    "coltouse=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(Group_type))\n",
    "#options(repr.plot.width=12, repr.plot.height=5)\n",
    "ggplot() +\n",
    "geom_bar(aes(y = count, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"count\") + labs(y = \"count of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = count, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"stack\") +\n",
    "scale_y_continuous() +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"count\") + labs(y = \"count of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_celltype, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by cell type\") + labs(y = paste0(\"pecentage of cells in given batch's \",tag_celltype),fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_all, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"fill\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by all cells\") + labs(y = \"pecentage of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "\n",
    "ggplot() +\n",
    "geom_bar(aes(y = percent_by_all, x = patient, fill = factor(majorcelltype,levels=Group_type)), data = tmp,stat=\"identity\",position = \"stack\") +\n",
    "scale_y_continuous(labels = percent_format()) +\n",
    "theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "facet_grid( ~ tissue,scales=\"free_x\", space=\"free_x\") + \n",
    "ggtitle(\"percent by all cells\") + labs(y = \"pecentage of cells in given batch\",fill='major cell types') + \n",
    "scale_fill_manual(\"legend\", values = coltouse) \n",
    "dev.off()\n",
    "\n",
    "# define the summary function\n",
    "f <- function(x) {\n",
    "  r <- quantile(x, probs = c(0.1, 0.25, 0.5, 0.75, 0.9))\n",
    "  names(r) <- c(\"ymin\", \"lower\", \"middle\", \"upper\", \"ymax\")\n",
    "  r\n",
    "}\n",
    "\n",
    "o <- function(x) {\n",
    "  subset(x, x < quantile(x)[2] | quantile(x)[4] < x)\n",
    "}\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(3,\"Set1\")\n",
    "\n",
    "options(warn=-1) #turn off warning\n",
    "pdf(paste0(\"healthy vs IBD in \",tag_celltype,\" subtype.pdf\"),width=25,height=12)\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=count, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"count of cells in given batch\") + ggtitle(\"count\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=percent_by_celltype, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"pecentage of cells in given batch\") + ggtitle(\"percent by cell type\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(4,\"Set1\")\n",
    "ggplot(tmp, aes(x=disease_status, y=percent_by_all, fill=disease_status)) + \n",
    "facet_wrap( ~ tissue + majorcelltype,scales=\"free\",nrow=3,labeller = label_wrap_gen(multi_line=FALSE)) + \n",
    "stat_summary(fun.data=f, geom=\"boxplot\") +\n",
    "stat_summary(fun.y=o, geom=\"point\") +\n",
    "labs(y = \"pecentage of cells in given batch\") + ggtitle(\"percent by all cells\") +\n",
    "stat_compare_means(label.x = 1.5,comparisons = list(c(\"Healthy\",\"IBD\"))) + #aes(label = ..p.signif..),\n",
    "scale_fill_manual(values=coltouse[2:1])\n",
    "dev.off()\n",
    "options(warn=0) #turn on warning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### gene-group heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#contrust matrix, each row is a state\n",
    "ExpressionMat_by_gp=cbind(Exp_Seurat@ident,2^t(Exp_Seurat@data))\n",
    "colnames(ExpressionMat_by_gp)[1]=\"states\"\n",
    "tagV=c(\"TRAV\",\"TRAC\",\"TRBV\",\"TRBC\",\"TRGV\",\"TRGC\",\"TRDV\",\"TRDC\",\n",
    "       \"IGHV\",\"IGHG\",\"IGHA\",\"IGHD\",\"IGHM\",\"IGHE\",\"IGLV\",\"IGLL\",\"IGKV\",\"IGLC\",\"IGKC\")\n",
    "togetrid=c(\"IGHMBP2\")\n",
    "for(i in 1:length(tagV)){\n",
    "    tagG=grep(paste0(\"^\",tagV[i]),colnames(ExpressionMat_by_gp))\n",
    "    tagG=setdiff(tagG,togetrid)\n",
    "    if(length(tagG)==0){\n",
    "        next\n",
    "    }\n",
    "    tmp=ExpressionMat_by_gp[,tagG]\n",
    "    if(!is.null(ncol(tmp))){\n",
    "        tmp=rowSums(tmp)\n",
    "    }\n",
    "    ExpressionMat_by_gp=cbind(ExpressionMat_by_gp,tmp)\n",
    "    colnames(ExpressionMat_by_gp)[ncol(ExpressionMat_by_gp)]=tagV[i]\n",
    "}\n",
    "ExpressionMat_by_gp=data.table(ExpressionMat_by_gp)\n",
    "ExpressionMat_by_gp=ExpressionMat_by_gp[,lapply(.SD, mean), by=states]\n",
    "ExpressionMat_by_gp=ExpressionMat_by_gp[order(ExpressionMat_by_gp$states)]\n",
    "ExpressionMat_by_gp[,states:=NULL]\n",
    "ExpressionMat_by_gp=data.frame(ExpressionMat_by_gp)\n",
    "colnames(ExpressionMat_by_gp)=gsub(\"\\\\.\",\"-\",colnames(ExpressionMat_by_gp))\n",
    "ExpressionMat_by_gp=log2(ExpressionMat_by_gp)\n",
    "rownames(ExpressionMat_by_gp)=paste0(\"gp_\",levels(Exp_Seurat@ident))\n",
    "write.table(t(ExpressionMat_by_gp),paste0(\"mediam_gene_expression_\",tag_celltype,\".tsv\"),sep=\"\\t\",col.names=T,row.names=T,quote=F)\n",
    "saveRDS(ExpressionMat_by_gp,file=paste0(\"ExpressionMat_by_gp_\",tag_celltype,\".Rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "B3GALT1, NCR2, IL22, RORC, KIR2DL1, KIR2DL2, KIRDL5A, KIR2DL5B, KIR3DL3, CEACAM1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ExpressionMat_by_gp=readRDS(paste0(\"ExpressionMat_by_gp_\",tag_celltype,\".Rds\"))\n",
    "\n",
    "#NK GENE\n",
    "taggene=c(\"NCAM1\",\"SELPLG\",\"KLRG1\",\"B3GALT1\",\"KLRD1\",\"SELL\",\"FCGR3A\",\"NCR1\",\"NCR3\",\"NCR2\",\"IL22\",\"IFNG\",\"RORC\",\"NFIL3\",\"KLRK1\",\n",
    "          \"KIR2DL1\",\"KIR2DL2\",\"KIR2DL3\",\"KIRDL5A\",\"KIR2DL5B\",\"KIR3DL1\",\"KIR3DL2\",\"KIR3DL3\",\"TYROBP\",\"HCST\",\"CD160\",\"CD226\",\n",
    "          \"CRTAM\",\"CD27\",\"SEMA4D\",\"CD96\",\"CEACAM1\",\"SLAMF7\",\"CD244\",\"SLAMf6\")\n",
    "\n",
    "taggene=toupper(taggene)\n",
    "tagExpressionNormed=ExpressionMat_by_gp[,which(colnames(ExpressionMat_by_gp) %in% taggene)]\n",
    "missing_gene=paste0(taggene[!taggene %in% colnames(ExpressionMat_by_gp)],collapse = \", \")\n",
    "message(missing_gene)\n",
    "\n",
    "#rownames(tagExpressionNormed)=NKcell_assignment_expanded$subcelltype[order(as.numeric(NKcell_assignment_expanded$cluster))]\n",
    "\n",
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "pdf(paste0(\"clusterGeneExp_\",tag_celltype,\"_heatmap.pdf\"),width=15, height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "heatmap.2(scale(tagExpressionNormed),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"cell assignment vs. gene\\nmissing:\",missing_gene),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "\n",
    "heatmap.2(as.matrix(tagExpressionNormed),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"cell assignment vs. gene\\nmissing:\",missing_gene),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### VDJ vs RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0(\"VDJassignment_vs_RNAassignment_\",tag_celltype,\"_heatmap.pdf\"),width=6,height=9)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "tmp=table(Exp_Seurat@ident,Exp_Seurat@meta.data$cell_assignment)\n",
    "\n",
    "heatmap.2(t(apply(tmp,1,function(x){x/sum(x)})),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=F,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"RNA-based cell type assignment\\n vs. VDJ based\"),key.title=\"\",key.xlab=\"\",margins=c(15,10),cellnote=tmp)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0(\"TSNE_mito_\",tag_celltype,\".pdf\"),width=18,height=15)\n",
    "par(mfrow=c(3,3))\n",
    "plotByGroup(Exp_Seurat,Group_assignment=sample_list[Exp_Seurat@meta.data$sample_gp],\n",
    "            Group_type=sample_list,main=\"Batchs\") #batchs\n",
    "plotByGroup(Exp_Seurat,Group_assignment=tissue_type[Exp_Seurat@meta.data$tissue_assignment],\n",
    "            Group_type=tissue_type,main=\"tissue type\") #tissue\n",
    "plotByGroup(Exp_Seurat,Group_assignment=patient_type[Exp_Seurat@meta.data$patient_assignment],\n",
    "            Group_type=patient_type,main=\"patient id\") #patient id\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@meta.data$cell_assignment,\n",
    "            Group_type=c('failed BCR/TCR','B cell','T cell','dual-label','unknown'),main=\"B/T classification based on VDJ\",backGroundGroup=\"unknown\") #BT\n",
    "plotByGroup(Exp_Seurat,Group_assignment=subcelltype_assignment,Group_type=celltype_assignment_expanded$subcelltype\n",
    "            ,main=\"rough cluster\",Addtext=T)\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@ident,main=\"NK cluster\",Addtext=T)\n",
    "plotByGroup(Exp_Seurat,Group_assignment=c(\"healthy\",\"diseased\")[Exp_Seurat@meta.data$disease_assignment],\n",
    "            Group_type=c(\"diseased\",\"healthy\"),main=\"disease condition\") #patient id\n",
    "#plotByScore(Exp_Seurat,Score_assignment=Exp_Seurat@meta.data$S.Score,main=\"\") \n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "clusterVsPatient=table(Exp_Seurat@ident,Exp_Seurat@meta.data$sample_gp)\n",
    "colnames(clusterVsPatient)=sample_list[as.numeric(colnames(clusterVsPatient))]\n",
    "clusterVsPatient_colPercent=round(apply(clusterVsPatient,2,function(x){x/sum(x)}),2)\n",
    "clusterVsPatient_rowPercent=round(t(apply(clusterVsPatient,1,function(x){x/sum(x)})),2)\n",
    "write.table(clusterVsPatient,file=paste0(\"clusterVsPatient_table_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(clusterVsPatient_rowPercent,file=paste0(\"clusterVsPatient_ClusterPercentage_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(clusterVsPatient_colPercent,file=paste0(\"clusterVsPatient_PatientPercentage_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "clusterVsPatient_rowPercent[clusterVsPatient_rowPercent>0.2]=0.2\n",
    "clusterVsPatient_colPercent[clusterVsPatient_colPercent>0.5]=0.5\n",
    "\n",
    "pdf(paste0(\"clusterSample_\",tag_celltype,\"_heatmap.pdf\"),width=15, height=8)\n",
    "heatmap.2(clusterVsPatient_colPercent,col=colorRampPalette(col2),\n",
    "          lhei=c(1,5),lwid=c(0.1,5),margins = c(10,10),trace=\"none\",density.info=\"none\",dendrogram='col',Colv=T,Rowv=T,\n",
    "          notecol=\"black\",key=F,cellnote=clusterVsPatient)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### heatmap cell-DEG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in `[<-.factor`(`*tmp*`, tag_unit, value = c(1L, 1L, 1L, 1L, 1L, :\n",
      "“invalid factor level, NA generated”"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "Group_type=unique(Exp_Seurat@ident)\n",
    "Group_type=Group_type[gtools::mixedorder(Group_type)]\n",
    "\n",
    "marker_genes=readRDS(paste0(\"markerGenes_\",base_Rdata_name,\"_\",tag_celltype,\".rds\"))\n",
    "get5k=Subsample_by_group(Exp_Seurat@ident,3000)\n",
    "toplot=ExpressionMat_magic[get5k,unique(marker_genes$gene)] #t(Exp_Seurat@data[unique(marker_genes$gene),get5k])\n",
    "\n",
    "tmp=as.numeric(gp_name_replacing(Exp_Seurat@ident[get5k],Group_type,1:length(Group_type)))\n",
    "globall_ordering=fastcluster::hclust.vector(toplot,method='ward')\n",
    "globall_sorting_idx=match(1:length(get5k),globall_ordering$order)\n",
    "colOrder=order(tmp,Exp_Seurat@meta.data$disease_assignment[get5k],globall_sorting_idx)\n",
    "\n",
    "toplot=toplot[colOrder,]\n",
    "toplot=t(scale(toplot))\n",
    "toplot[toplot>1.5]=1.5\n",
    "toplot[toplot<(-1.5)]=-1.5\n",
    "\n",
    "ColforDisease=RColorBrewer::brewer.pal(3,\"Set1\")[2:1]\n",
    "ColforSubCelltype=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(Group_type))\n",
    "ColforPatient_ID=colorRampPalette(RColorBrewer::brewer.pal(9,\"Set1\"))(length(patient_type))\n",
    "ColSideColors=cbind(patient_ID=ColforPatient_ID[Exp_Seurat@meta.data$patient_assignment],\n",
    "                    disease=ColforDisease[Exp_Seurat@meta.data$disease_assignment],\n",
    "                    subcelltype=ColforSubCelltype[match(Exp_Seurat@ident,Group_type)])\n",
    "ColSideColors=ColSideColors[get5k[colOrder],]\n",
    "\n",
    "RowSideColors=matrix(\"white\",nrow(toplot),max(as.numeric(marker_genes$cluster)))\n",
    "colnames(RowSideColors)=Group_type\n",
    "for(i in 1:max(as.numeric(marker_genes$cluster))){\n",
    "    tagG=marker_genes$gene[marker_genes$cluster==i-1]\n",
    "    tagsubtype=Group_type[i]\n",
    "    RowSideColors[rownames(toplot) %in% tagG,which(Group_type==tagsubtype)]=ColforSubCelltype[which(Group_type==tagsubtype)]\n",
    "}\n",
    "\n",
    "#ColSideColors can be a data.frame with each col indicate one color assignment\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "pdf(file=paste0(tag_celltype,\"_DEGbycluster_heatmap.pdf\"),width=15,height=10)\n",
    "heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,RowSideColors=RowSideColors,ColSideColors=ColSideColors,\n",
    "                   labRow=F,labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none')\n",
    "dev.off()\n",
    "\n",
    "png(file=paste0(tag_celltype,\"_DEGbycluster_heatmap.png\"),width=15,height=10,units = \"in\",res=150)\n",
    "heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,RowSideColors=RowSideColors,ColSideColors=ColSideColors,\n",
    "                   labRow=F,labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none')\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "##### PER GENE plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "require(feather)\n",
    "#MAGIC\n",
    "forMagic=\"For_magic.feather\"\n",
    "afterMagic=\"Magic_out.feather\"\n",
    "\n",
    "temp_Exp=as.data.frame(t(Exp_Seurat@data))\n",
    "#magic\n",
    "write_feather(temp_Exp, forMagic)\n",
    "system(paste0(\"python /home/ahe/tools/Lightbulb/src/Lightbulb_MagicWrapper.py --matx \",forMagic,\" --out \",afterMagic))\n",
    "ExpressionMat_magic = data.matrix(read_feather(afterMagic))\n",
    "colnames(ExpressionMat_magic)=gsub(\"MAGIC \", \"\", colnames(ExpressionMat_magic))\n",
    "\n",
    "saveRDS(ExpressionMat_magic,file=paste0(tag_celltype,\"_magic.rds\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'B3GALT1'</li>\n",
       "\t<li>'NCR2'</li>\n",
       "\t<li>'IL22'</li>\n",
       "\t<li>'RORC'</li>\n",
       "\t<li>'KIR2DL1'</li>\n",
       "\t<li>'KIR2DL2'</li>\n",
       "\t<li>'KIRDL5A'</li>\n",
       "\t<li>'KIR2DL5B'</li>\n",
       "\t<li>'KIR3DL3'</li>\n",
       "\t<li>'CEACAM1'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B3GALT1'\n",
       "\\item 'NCR2'\n",
       "\\item 'IL22'\n",
       "\\item 'RORC'\n",
       "\\item 'KIR2DL1'\n",
       "\\item 'KIR2DL2'\n",
       "\\item 'KIRDL5A'\n",
       "\\item 'KIR2DL5B'\n",
       "\\item 'KIR3DL3'\n",
       "\\item 'CEACAM1'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B3GALT1'\n",
       "2. 'NCR2'\n",
       "3. 'IL22'\n",
       "4. 'RORC'\n",
       "5. 'KIR2DL1'\n",
       "6. 'KIR2DL2'\n",
       "7. 'KIRDL5A'\n",
       "8. 'KIR2DL5B'\n",
       "9. 'KIR3DL3'\n",
       "10. 'CEACAM1'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"B3GALT1\"  \"NCR2\"     \"IL22\"     \"RORC\"     \"KIR2DL1\"  \"KIR2DL2\" \n",
       " [7] \"KIRDL5A\"  \"KIR2DL5B\" \"KIR3DL3\"  \"CEACAM1\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ExpressionMat_magic=readRDS(paste0(tag_celltype,\"_magic.rds\"))\n",
    "taggene=c(\"NCAM1\",\"SELPLG\",\"KLRG1\",\"B3GALT1\",\"KLRD1\",\"SELL\",\"FCGR3A\",\"NCR1\",\"NCR3\",\"NCR2\",\"IL22\",\"IFNG\",\"RORC\",\"NFIL3\",\"KLRK1\",\n",
    "          \"KIR2DL1\",\"KIR2DL2\",\"KIR2DL3\",\"KIRDL5A\",\"KIR2DL5B\",\"KIR3DL1\",\"KIR3DL2\",\"KIR3DL3\",\"TYROBP\",\"HCST\",\"CD160\",\"CD226\",\n",
    "          \"CRTAM\",\"CD27\",\"SEMA4D\",\"CD96\",\"CEACAM1\",\"SLAMF7\",\"CD244\",\"SLAMf6\")\n",
    "\n",
    "taggene=toupper(taggene)\n",
    "tagExpressionNormed=ExpressionMat_magic[,which(colnames(ExpressionMat_magic) %in% taggene)]\n",
    "tagExpressionNormed=tagExpressionNormed[,order(colnames(tagExpressionNormed))]\n",
    "taggene[!taggene %in% colnames(ExpressionMat_magic)]\n",
    "\n",
    "rbPal <- colorRampPalette(c('gray','#2b83ba','#abdda4','#ffffbf','#fdae61','#d7191c'))\n",
    "selected=sample(1:nrow(tagExpressionNormed),2000)\n",
    "\n",
    "pdf(paste0(tag_celltype,\"_per gene mito TSNE_magic.pdf\"),width=22,height=15)\n",
    "par(mfrow=c(4,6))\n",
    "for(i in 1:ncol(tagExpressionNormed)){\n",
    "    currentid=colnames(tagExpressionNormed)[i]\n",
    "    #calculate breaks\n",
    "    lowerbound=sort(tagExpressionNormed[selected,i])[round(length(selected)/33)]\n",
    "    upperbound=sort(tagExpressionNormed[selected,i])[round(length(selected)/33*32)]\n",
    "    if(upperbound!=lowerbound){\n",
    "        breaks=c(min(tagExpressionNormed[selected,i])-1,seq(lowerbound, upperbound, length.out=25),max(tagExpressionNormed[selected,i])+1)\n",
    "        cols=rbPal(25)[as.numeric(cut(tagExpressionNormed[selected,i],breaks = breaks, include.lowest=TRUE))]\n",
    "    }else{\n",
    "        breaks=c(min(tagExpressionNormed[selected,i])-1,max(tagExpressionNormed[selected,i])/2,max(tagExpressionNormed[selected,i]))\n",
    "        cols=rbPal(2)[as.numeric(cut(tagExpressionNormed[selected,i],breaks = breaks, include.lowest=TRUE))]\n",
    "    }\n",
    "    plot(Exp_Seurat@dr$tsne@cell.embeddings[selected,],pch=19,cex=0.35,col=cols,main=currentid)  \n",
    "}\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### heatmap of DEG between healthy and diseas in each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "DEG=readRDS(paste0(\"DE_healthyVSdisease_within_cluster_\",tag_celltype,\".Rds\"))\n",
    "ExpressionMat_magic=readRDS(file=paste0(tag_celltype,\"_magic.rds\"))\n",
    "\n",
    "#new naming\n",
    "names(DEG)=gsub(\"subcluster_\",\"NK_\",names(DEG))\n",
    "\n",
    "coltouse=RColorBrewer::brewer.pal(3,\"Set1\")\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "\n",
    "for(i in 1:length(DEG)){\n",
    "    if(sum(DEG[[i]]$DE)>1){\n",
    "        tagG=rownames(DEG[[i]])[DEG[[i]]$DE]\n",
    "        taghealthy=ExpressionMat_magic[paste0(\"NK_\",Exp_Seurat@ident)==names(DEG)[i] & Exp_Seurat@meta.data$disease_assignment==1,tagG]\n",
    "        tmp=fastcluster::hclust.vector(taghealthy,method='single')\n",
    "        taghealthy=taghealthy[tmp$order,]\n",
    "        tagIBD=ExpressionMat_magic[paste0(\"NK_\",Exp_Seurat@ident)==names(DEG)[i] & Exp_Seurat@meta.data$disease_assignment==2,tagG]\n",
    "        tmp=fastcluster::hclust.vector(tagIBD,method='single')\n",
    "        tagIBD=tagIBD[tmp$order,]\n",
    "        colcor=c(rep(coltouse[2],nrow(taghealthy)),rep(coltouse[1],nrow(tagIBD)))\n",
    "        toplot=rbind(taghealthy,tagIBD)\n",
    "        toplot=t(scale(toplot))\n",
    "        toplot[toplot>1.5]=1.5\n",
    "        toplot[toplot<(-1.5)]=-1.5\n",
    "        pdf(paste0(\"heatmap_DEG_healthyVSdiseased_\",names(DEG)[i],\".pdf\"))\n",
    "        heatmap3::heatmap3(toplot,col=colorRampPalette(col2)(50),useRaster=F,ColSideColors=colcor,ColSideLabs=\"IBD Status\",\n",
    "                   labRow=rownames(toplot),labCol=F,showColDendro=F,showRowDendro=F,Colv=NA,scale='none',main=names(DEG)[i])\n",
    "        dev.off()\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MPhi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_macrophage.Rds\")\n",
    "tag_celltype=\"Mphi\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#cell type assignment\n",
    "celltype_assignment_df=data.frame(cluster=c(\"17,16\",\"8,2\",\"13,5\",\"6,0\",\"7,10\",\"11,1\",\"3\",\"18,12,9,19\",\"21,20\",\"4,14,15\"),\n",
    "                                  celltype=c(\"Naive_B\",\"Memory_B\",\"Plasma\",\"NKT\",\"CD8\",\"CD4\",\"gdT\",\"NK\",\"unknown\",\"M_phi\"),\n",
    "                                  stringsAsFactors=F)\n",
    "\n",
    "celltype_assignment_expanded=cellTypeDF_processing(celltype_assignment_df)\n",
    "\n",
    "celltype_assignment=gp_name_replacing(as.character(Exp_Seurat@meta.data$global_gps),celltype_assignment_expanded$cluster,celltype_assignment_expanded$celltype)\n",
    "subcelltype_assignment=gp_name_replacing(as.character(Exp_Seurat@meta.data$global_gps),celltype_assignment_expanded$cluster,celltype_assignment_expanded$subcelltype)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {},
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#contrust matrix, each row is a state\n",
    "ExpressionMat_by_gp=cbind(Exp_Seurat@ident,2^t(Exp_Seurat@data))\n",
    "colnames(ExpressionMat_by_gp)[1]=\"states\"\n",
    "tagV=c(\"TRAV\",\"TRAC\",\"TRBV\",\"TRBC\",\"TRGV\",\"TRGC\",\"TRDV\",\"TRDC\",\n",
    "       \"IGHV\",\"IGHG\",\"IGHA\",\"IGHD\",\"IGHM\",\"IGHE\",\"IGLV\",\"IGLL\",\"IGKV\",\"IGLC\",\"IGKC\")\n",
    "togetrid=c(\"IGHMBP2\")\n",
    "for(i in 1:length(tagV)){\n",
    "    tagG=grep(paste0(\"^\",tagV[i]),colnames(ExpressionMat_by_gp))\n",
    "    tagG=setdiff(tagG,togetrid)\n",
    "    if(length(tagG)==0){\n",
    "        next\n",
    "    }\n",
    "    tmp=ExpressionMat_by_gp[,tagG]\n",
    "    if(!is.null(ncol(tmp))){\n",
    "        tmp=rowSums(tmp)\n",
    "    }\n",
    "    ExpressionMat_by_gp=cbind(ExpressionMat_by_gp,tmp)\n",
    "    colnames(ExpressionMat_by_gp)[ncol(ExpressionMat_by_gp)]=tagV[i]\n",
    "}\n",
    "ExpressionMat_by_gp=data.table(ExpressionMat_by_gp)\n",
    "ExpressionMat_by_gp=ExpressionMat_by_gp[,lapply(.SD, mean), by=states]\n",
    "ExpressionMat_by_gp=ExpressionMat_by_gp[order(ExpressionMat_by_gp$states)]\n",
    "ExpressionMat_by_gp[,states:=NULL]\n",
    "ExpressionMat_by_gp=data.frame(ExpressionMat_by_gp)\n",
    "colnames(ExpressionMat_by_gp)=gsub(\"\\\\.\",\"-\",colnames(ExpressionMat_by_gp))\n",
    "ExpressionMat_by_gp=log2(ExpressionMat_by_gp)\n",
    "rownames(ExpressionMat_by_gp)=paste0(\"gp_\",levels(Exp_Seurat@ident))\n",
    "write.table(t(ExpressionMat_by_gp),paste0(\"mediam_gene_expression_\",tag_celltype,\".tsv\"),sep=\"\\t\",col.names=T,row.names=T,quote=F)\n",
    "saveRDS(ExpressionMat_by_gp,file=paste0(\"ExpressionMat_by_gp_\",tag_celltype,\".Rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ExpressionMat_by_gp=readRDS(paste0(\"ExpressionMat_by_gp_\",tag_celltype,\".Rds\"))\n",
    "\n",
    "#MPhi GENE\n",
    "taggene=c()\n",
    "\n",
    "taggene=toupper(taggene)\n",
    "tagExpressionNormed=ExpressionMat_by_gp[,which(colnames(ExpressionMat_by_gp) %in% taggene)]\n",
    "taggene[!taggene %in% colnames(ExpressionMat_by_gp)]\n",
    "\n",
    "rownames(tagExpressionNormed)=MPhicell_assignment_expanded$subcelltype[order(as.numeric(MPhicell_assignment_expanded$cluster))]\n",
    "\n",
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "\n",
    "pdf(paste0(\"clusterGeneExp_\",tag_celltype,\"_heatmap.pdf\"),width=15, height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "heatmap.2(scale(tagExpressionNormed),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"cell assignment vs. gene\"),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "\n",
    "heatmap.2(as.matrix(tagExpressionNormed),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"cell assignment vs. gene\"),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### VDJ vs RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0(\"VDJassignment_vs_RNAassignment_\",tag_celltype,\"_heatmap.pdf\"),width=6,height=9)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "tmp=table(Exp_Seurat@ident,Exp_Seurat@meta.data$cell_assignment)\n",
    "\n",
    "heatmap.2(t(apply(tmp,1,function(x){x/sum(x)})),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=T,Rowv=F,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=paste0(\"RNA-based cell type assignment\\n vs. VDJ based\"),key.title=\"\",key.xlab=\"\",margins=c(15,10),cellnote=tmp)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0(\"TSNE_mito_\",tag_celltype,\".pdf\"),width=18,height=15)\n",
    "par(mfrow=c(3,3))\n",
    "plotByGroup(Exp_Seurat,Group_assignment=sample_list[Exp_Seurat@meta.data$sample_gp],\n",
    "            Group_type=sample_list,main=\"Batchs\") #batchs\n",
    "plotByGroup(Exp_Seurat,Group_assignment=tissue_type[Exp_Seurat@meta.data$tissue_assignment],\n",
    "            Group_type=tissue_type,main=\"tissue type\") #tissue\n",
    "plotByGroup(Exp_Seurat,Group_assignment=patient_type[Exp_Seurat@meta.data$patient_assignment],\n",
    "            Group_type=patient_type,main=\"patient id\") #patient id\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@meta.data$cell_assignment,\n",
    "            Group_type=c('failed BCR/TCR','B cell','T cell','dual-label','unknown'),main=\"B/T classification based on VDJ\",backGroundGroup=\"unknown\") #BT\n",
    "plotByGroup(Exp_Seurat,Group_assignment=subcelltype_assignment,Group_type=celltype_assignment_expanded$subcelltype\n",
    "            ,main=\"rough cluster\",Addtext=T)\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@ident,main=\"NK cluster\",Addtext=T)\n",
    "plotByGroup(Exp_Seurat,Group_assignment=c(\"healthy\",\"diseased\")[Exp_Seurat@meta.data$disease_assignment],\n",
    "            Group_type=c(\"diseased\",\"healthy\"),main=\"disease condition\") #patient id\n",
    "#plotByScore(Exp_Seurat,Score_assignment=Exp_Seurat@meta.data$S.Score,main=\"\") \n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "clusterVsPatient=table(Exp_Seurat@ident,Exp_Seurat@meta.data$sample_gp)\n",
    "colnames(clusterVsPatient)=sample_list[as.numeric(colnames(clusterVsPatient))]\n",
    "clusterVsPatient_colPercent=round(apply(clusterVsPatient,2,function(x){x/sum(x)}),2)\n",
    "clusterVsPatient_rowPercent=round(t(apply(clusterVsPatient,1,function(x){x/sum(x)})),2)\n",
    "write.table(clusterVsPatient,file=paste0(\"clusterVsPatient_table_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(clusterVsPatient_rowPercent,file=paste0(\"clusterVsPatient_ClusterPercentage_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "write.table(clusterVsPatient_colPercent,file=paste0(\"clusterVsPatient_PatientPercentage_\",tag_celltype,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "clusterVsPatient_rowPercent[clusterVsPatient_rowPercent>0.2]=0.2\n",
    "clusterVsPatient_colPercent[clusterVsPatient_colPercent>0.5]=0.5\n",
    "\n",
    "pdf(paste0(\"clusterSample_\",tag_celltype,\"_heatmap.pdf\"),width=15, height=8)\n",
    "heatmap.2(clusterVsPatient_colPercent,col=colorRampPalette(col2),\n",
    "          lhei=c(1,5),lwid=c(0.1,5),margins = c(10,10),trace=\"none\",density.info=\"none\",dendrogram='col',Colv=T,Rowv=T,\n",
    "          notecol=\"black\",key=F,cellnote=clusterVsPatient)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## define all celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#T cell\n",
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_Tcell.Rds\")\n",
    "Tcell_assignment_df=data.frame(cluster=c(\"12,9,5,15\",\"6,16\",\"11\",\"1,0,10,7\",\"14\",\"4,17,3,8,2,13\"),\n",
    "                               celltype=c(\"CD8_TEM_effector\",\"CD8\",\"CD8_TCM\",\"NKT\",\"gdT\",\"CD4\"),stringsAsFactors=F)\n",
    "Tcell_assignment_expanded=cellTypeDF_processing(Tcell_assignment_df)\n",
    "Tcellsubcelltype_assignment=gp_name_replacing(as.character(Exp_Seurat@ident),Tcell_assignment_expanded$cluster,Tcell_assignment_expanded$subcelltype)\n",
    "\n",
    "cell_assignment=data.frame(cell_id=colnames(Exp_Seurat@data),gp_assign=Tcellsubcelltype_assignment)\n",
    "\n",
    "#B cell\n",
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_Bcell.Rds\")\n",
    "Bcell_assignment_df=data.frame(cluster=c(\"6,14,5,2,13,10,12\",\"3,0,1,8\",\"11\",\"4,7,9\"),\n",
    "                                  celltype=c(\"Plasma\",\"Memory_B\",\"Breg\",\"Naive_B\"),\n",
    "                                  stringsAsFactors=F)\n",
    "Bcell_assignment_expanded=cellTypeDF_processing(Bcell_assignment_df)\n",
    "Bcellsubcelltype_assignment=gp_name_replacing(as.character(Exp_Seurat@ident),Bcell_assignment_expanded$cluster,Bcell_assignment_expanded$subcelltype)\n",
    "\n",
    "cell_assignment=rbind(data.frame(cell_id=colnames(Exp_Seurat@data),gp_assign=Bcellsubcelltype_assignment),cell_assignment)\n",
    "\n",
    "#NK\n",
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_NKcell.Rds\")\n",
    "cell_assignment=rbind(data.frame(cell_id=colnames(Exp_Seurat@data),gp_assign=paste0(\"NK_\",Exp_Seurat@ident)),cell_assignment)\n",
    "\n",
    "#Mphi\n",
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_macrophage.Rds\")\n",
    "cell_assignment=rbind(data.frame(cell_id=colnames(Exp_Seurat@data),gp_assign=paste0(\"Mphi_\",Exp_Seurat@ident)),cell_assignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### loading Exp_Seurat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# All\n",
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_UNKNOWNrm.Rds\")\n",
    "out_prefix=\"All\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# healthy I,R,P\n",
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_healthyIRP.Rds\")\n",
    "out_prefix=\"HealthyIRP\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Disease R,P\n",
    "Exp_Seurat=readRDS(file=\"Exp_Seurat_DiseaseRP.Rds\")\n",
    "out_prefix=\"DiseaseRP\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#cell type assignment\n",
    "rough_assignment_df=data.frame(cluster=c(\"17,16\",\"8,2\",\"13,5\",\"6,0\",\"7,10\",\"11,1\",\"3\",\"18,12,9,19\",\"21,20\",\"4,14,15\"),\n",
    "                                  celltype=c(\"Naive_B\",\"Memory_B\",\"Plasma\",\"NKT\",\"CD8\",\"CD4\",\"gdT\",\"NK\",\"unknown\",\"Mphi\"),\n",
    "                                  stringsAsFactors=F)\n",
    "\n",
    "rough_assignment_expanded=cellTypeDF_processing(rough_assignment_df)\n",
    "rough_assignment=gp_name_replacing(as.character(Exp_Seurat@ident),rough_assignment_expanded$cluster,rough_assignment_expanded$celltype)\n",
    "rough_assignment_majorgroup=gp_name_replacing(rough_assignment,c(\"Naive_B\",\"Memory_B\",\"Plasma\",\"NKT\",\"CD8\",\"CD4\",\"gdT\"),c(\"B\",\"B\",\"B\",\"T\",\"T\",\"T\",\"T\"))\n",
    "specific_assignment_subgroup=cell_assignment$gp_assign[match(colnames(Exp_Seurat@data),cell_assignment$cell_id)]\n",
    "\n",
    "specific_assignment_subtype=gsub(\"_[0-9]+\",\"\",specific_assignment_subgroup)\n",
    "specific_assignment_celltype=gp_name_replacing(specific_assignment_subtype,c(\"CD8_TEM_effector\",\"CD8_TCM\"),c(\"CD8\",\"CD8\"),force_replace=T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "clusterVsSample=table(specific_assignment_celltype,Exp_Seurat@meta.data$sample_gp)\n",
    "colnames(clusterVsSample)=sample_list[as.numeric(colnames(clusterVsSample))]\n",
    "clusterVsSample_colPercent=round(apply(clusterVsSample,2,function(x){x/sum(x)}),2)\n",
    "clusterVsSample_rowPercent=round(t(apply(clusterVsSample,1,function(x){x/sum(x)})),2)\n",
    "write.table(clusterVsSample,file=paste0(\"clusterVsSample_table_\",out_prefix,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "#write.table(clusterVsSample_colPercent,file=paste0(\"clusterVsSample_ClusterPercentage_\",out_prefix,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "#write.table(clusterVsSample_rowPercent,file=paste0(\"clusterVsSample_SamplePercentage_\",out_prefix,\".csv\"),sep=\",\",col.names=T,row.names = T,quote=F)\n",
    "\n",
    "pdf(paste0(\"clusterVsSample_\",out_prefix,\"_heatmap.pdf\"),width=15,height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "heatmap.2(clusterVsSample_colPercent,col=colorRampPalette(col2),\n",
    "          lhei=c(1,5),lwid=c(0.1,5),margins = c(10,10),trace=\"none\",density.info=\"none\",dendrogram='col',Colv=T,Rowv=F,\n",
    "          notecol=\"black\",key=F,cellnote=clusterVsSample)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(paste0(\"TSNE_mito_\",out_prefix,\".pdf\"),width=18,height=15)\n",
    "par(mfrow=c(3,3))\n",
    "plotByGroup(Exp_Seurat,Group_assignment=sample_list[Exp_Seurat@meta.data$sample_gp],\n",
    "            Group_type=sample_list,main=\"Batchs\") #batchs\n",
    "plotByGroup(Exp_Seurat,Group_assignment=tissue_type[Exp_Seurat@meta.data$tissue_assignment],\n",
    "            Group_type=tissue_type,main=\"tissue type\") #tissue\n",
    "plotByGroup(Exp_Seurat,Group_assignment=patient_type[Exp_Seurat@meta.data$patient_assignment],\n",
    "            Group_type=patient_type,main=\"patient id\") #patient id\n",
    "plotByGroup(Exp_Seurat,Group_assignment=Exp_Seurat@meta.data$cell_assignment,\n",
    "            Group_type=c('failed BCR/TCR','B cell','T cell','dual-label','unknown'),main=\"B/T classification based on VDJ\",backGroundGroup=\"unknown\") #BT\n",
    "plotByGroup(Exp_Seurat,Group_assignment=rough_assignment_majorgroup,Group_type=c(\"T\",\"B\",\"NK\",\"Mphi\")\n",
    "            ,main=\"major celltype\",Addtext=T)\n",
    "plotByGroup(Exp_Seurat,Group_assignment=specific_assignment_celltype,Group_type=c(\"CD8\",\"CD4\",\"NKT\",\"gdT\",\"Plasma\",\"Memory_B\",\"Breg\",\"Naive_B\",\"NK\",\"Mphi\")\n",
    "            ,main=\"sub celltype \",Addtext=T)\n",
    "plotByGroup(Exp_Seurat,Group_assignment=specific_assignment_subgroup,,main=\"specific subgroup\",Addtext=T)\n",
    "plotByGroup(Exp_Seurat,Group_assignment=c(\"healthy\",\"diseased\")[Exp_Seurat@meta.data$disease_assignment],\n",
    "            Group_type=c(\"healthy\",\"diseased\"),main=\"disease condition\") #patient id\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# old"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tag_celltype=c(\"macrophage\",\"NKcell\",\"Bcell\",\"Tcell\")\n",
    "base_Rdata_name=\"180607_RUVscale\"\n",
    "\n",
    "for(j in 1:length(tag_celltype)){\n",
    "    marker_genes=readRDS(paste0(\"markerGenes_\",base_Rdata_name,\"_\",tag_celltype[j],\".rds\"))\n",
    "    marker_gene_list=c()\n",
    "    for(i in 1:length(levels(Exp_Seurat@ident))){\n",
    "        marker_gene_list[[i]]=marker_genes$gene[marker_genes$cluster==levels(Exp_Seurat@ident)[i]]\n",
    "        names(marker_gene_list)[i]=paste0(\"cluster_\",levels(Exp_Seurat@ident)[i])\n",
    "    }\n",
    "    #marker_gene_list\n",
    "    write(paste0(tag_celltype[j],\"_marker:\"),file=paste0(\"markerGenes_\",tag_celltype[j],\".txt\"))\n",
    "    for(i in 1:length(marker_gene_list)){\n",
    "        write(paste0(names(marker_gene_list)[i],\":\"),file=paste0(\"markerGenes_\",tag_celltype[j],\".txt\"), append=T)\n",
    "        write(marker_gene_list[[i]],file=paste0(\"markerGenes_\",tag_celltype[j],\".txt\"), append=T,ncolumns=1000)\n",
    "        write(\"\",file=paste0(\"markerGenes_\",tag_celltype[j],\".txt\"), append=T)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# find marker genes\n",
    "# findMarkerGene_celltypeSpecific.R\n",
    "marker_genes=readRDS(paste0(\"markerGenes_\",base_Rdata_name,\"_\",tag_celltype,\".rds\"))\n",
    "\n",
    "marker_gene_list=c()\n",
    "for(i in 1:length(levels(Exp_Seurat@ident))){\n",
    "    marker_gene_list[[i]]=marker_genes$gene[marker_genes$cluster==levels(Exp_Seurat@ident)[i]]\n",
    "    names(marker_gene_list)[i]=paste0(\"cluster_\",levels(Exp_Seurat@ident)[i])\n",
    "}\n",
    "#marker_gene_list\n",
    "write(paste0(tag_celltype,\"_marker:\"),file=paste0(\"markerGenes_\",tag_celltype,\".txt\"))\n",
    "for(i in 1:length(marker_gene_list)){\n",
    "    write(paste0(names(marker_gene_list)[i],\":\"),file=paste0(\"markerGenes_\",tag_celltype,\".txt\"), append=T)\n",
    "    write(marker_gene_list[[i]],file=paste0(\"markerGenes_\",tag_celltype,\".txt\"), append=T,ncolumns=1000)\n",
    "    write(\"\",file=paste0(\"markerGenes_\",tag_celltype,\".txt\"), append=T)\n",
    "}\n",
    "\n",
    "DE_heatmap=t(Exp_Seurat@data)[,unique(marker_genes$gene)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "marker_gp_indicator=\n",
    "\n",
    "r <- raster(cur_mat)\n",
    "extent(r) <- extent(c(-180, 180, -90, 90))\n",
    "## Create a raster with the desired dimensions, and resample into it\n",
    "DE_heatmap_raster <- raster(nrow=2000, ncol=2000)\n",
    "DE_heatmap_raster= resample(r,DE_heatmap_raster)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fit=fastcluster::hclust.vector(DE_heatmap,method='single')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pdf(paste0(tag_celltype,\"_markerheatmap.pdf\"),width=8, height=8)\n",
    "col2=rev(RColorBrewer::brewer.pal(9,\"RdBu\"))\n",
    "heatmap.2(scale(DE_heatmap),col=colorRampPalette(col2),\n",
    "          trace=\"none\",density.info=\"none\",dendrogram='none',Colv=F,Rowv=T,notecol=\"black\",key=T,lhei=c(1,5),lwid=c(1,5),\n",
    "          main=aste0(tag_celltype,\" marker heatmap\"),key.title=\"\",key.xlab=\"\",margins=c(5,10))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "1376"
      ],
      "text/latex": [
       "1376"
      ],
      "text/markdown": [
       "1376"
      ],
      "text/plain": [
       "[1] 1376"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique(marker_genes$gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'IL17A'</li>\n",
       "\t<li>'RORC'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'IL17A'\n",
       "\\item 'RORC'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'IL17A'\n",
       "2. 'RORC'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"IL17A\" \"RORC\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>10363</li>\n",
       "\t<li>25173</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 10363\n",
       "\\item 25173\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 10363\n",
       "2. 25173\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 10363 25173"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dim(Exp_Seurat@data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'NKT_2'</li>\n",
       "\t<li>'NKT_2'</li>\n",
       "\t<li>'NKT_2'</li>\n",
       "\t<li>'NKT_2'</li>\n",
       "\t<li>'NKT_2'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'NKT\\_2'\n",
       "\\item 'NKT\\_2'\n",
       "\\item 'NKT\\_2'\n",
       "\\item 'NKT\\_2'\n",
       "\\item 'NKT\\_2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'NKT_2'\n",
       "2. 'NKT_2'\n",
       "3. 'NKT_2'\n",
       "4. 'NKT_2'\n",
       "5. 'NKT_2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"NKT_2\" \"NKT_2\" \"NKT_2\" \"NKT_2\" \"NKT_2\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "subcelltype_assignment[1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.4.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
